## Analysis of human lung pulmonary fibrosis data  using STExplorer
# The data used here are from
#
#
# ---------------------------------------------------------------------------- #
# ----Load packages ------------------------------------------------------------
library(STExplorer)

## Data manipulation/ plotting
library(ggplot2)
library(tidyverse)
library(patchwork)
library(ggpubr)
library(cols4all)
library(tidyquant)
library(ggdist)
library(ggthemes)

## Analysis
library(GSVA)
library(harmony)
library(Seurat)

# ---------------------------------------------------------------------------- #
# ----Prepare to load data -----------------------------------------------------
## Prepare vectors with the paths to the data folders
projectFolder <- "./../STExplorer_Analysis/Lung"
dataFolder <- paste0(projectFolder, "/data/hs_visium_spaceranger_output")
sampleDir <- list.dirs(dataFolder, recursive = FALSE)
sampleNames <- list.files(dataFolder)
sampleNames <- sampleNames[sampleNames != "README.txt"]
names(sampleDir) <- sampleNames


# ---------------------------------------------------------------------------- #
# ----Load Metadata ------------------------------------------------------------
## Load the metadata
metaDataFile <- paste0(projectFolder, "/data/hs_visium_metadata.tsv")
metadata <- read.table(metaDataFile, header = TRUE) %>%
  mutate(fibrotic_extent_score_by_pathologist_0.3 = as.character(fibrotic_extent_score_by_pathologist_0.3))
rownames(metadata) <- metadata$sample_id

## Load cell density-related information as generated by original authors
cellAbundanceFolder <- paste0(projectFolder, "/data/cell2location_habermann2020/")
cellGroupsFile <- paste0(projectFolder, "/data/hs_misc/habermann_cell_type_groups.csv")
cellGroups <- read.table(cellGroupsFile, header = TRUE, sep = ";")

## Load spot annotations
annotationFile <- paste0(projectFolder, "/data/hs_misc/hs_visium_merged_histo_annotations.csv")
annotation <- read.table(annotationFile, header = TRUE, sep = ",") %>%
  rename(Barcode = barcode, sample_id = sample_slide_id)

## Load gene annotations from biomart (as created by original authors)
annotation_gene <- read.table(file = file.path(projectFolder, "data/hs_misc", "hs_gene_biomart.2022-02-23.tsv"),
                              sep = "\t",
                              header = T,
                              stringsAsFactors = F) %>%
  rename(id = ensembl_gene_id, gene_name = hgnc_symbol)


# ---------------------------------------------------------------------------- #
# ----Select samples -----------------------------------------------------------
## Select samples for which there are annotations provided by the original authors
selection_samples <- names(sampleDir) %in% unique(annotation$sample_id)
sampleDir_selected <- sampleDir[selection_samples]
sampleNames_selected <- sampleNames[selection_samples]

## Update metadata to keep only the selected samples
metadata <- metadata[metadata$sample_id %in% sampleNames_selected,]


# ---------------------------------------------------------------------------- #
# ----Load data ----------------------------------------------------------------
## Create the MSFE object
msfe <- MetaSpatialFeatureExperiment()

## Import samples
for (i in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[i]
  message("Adding sample: ", sampleNames_selected[i])
  msfe <- addSFE(msfe,
                 read10xVisiumSFE(samples = sampleDir_selected[id],
                                  sample_id = sampleNames_selected[i],
                                  type = "HDF5",
                                  data = "filtered",
                                  images = "lowres",
                                  style = "W",
                                  zero.policy = TRUE))
}

## Import annotations
gTruth_list <- list()
for (id in sampleNames_selected) {
  gTruth_list[[id]] <- annotation[annotation$sample_id == id, c("Barcode", "sample_id", "annotation")]

}

## Import cell type abundances
cellAbundance <- vector(mode = "list", length = length(sampleNames_selected))
names(cellAbundance) <- sampleNames_selected
for (s in sampleNames_selected) {
  file <- paste0(cellAbundanceFolder, s, "_spot_cell_abundances_5pc.csv")
  df <- read.csv(file) %>%
    rename(Barcode = spot_id) %>%
    mutate(Barcode = gsub(paste0(s, "_"), "", Barcode),
           sample_id = s, .after = Barcode)
  cellAbundance[[s]] <- df
}


# ---------------------------------------------------------------------------- #
# ----Select genes -------------------------------------------------------------
## Get annotations
biomart <- createBiomart()
annotation_gene <- data.frame(gene_name = rowData(msfe@sfe_data[[1]])[["symbol"]])
rownames(annotation_gene) <- rownames(msfe@sfe_data[[1]])
annotation_gene <- annotateDataFrame(annotation_gene,
                                     biomart = biomart,
                                     add = c("biotype", "chromosome", "description"))

## Senescence Markers
senescenceMarkers <- c("CDKN2A", "CDKN1A", "GLB1", "LMNB1")
senescenceMarkers_ENSG <- setNames(
  annotation_gene$id[annotation_gene$gene_name %in% senescenceMarkers],
  annotation_gene$gene_name[annotation_gene$gene_name %in% senescenceMarkers]
)

## Genes Related to senescence and fibrosis
otherGenes <- c("IL6", "IL6R", "IL6ST", "GDF15", "CCL2", "TGFB1", "FAP", "VIM",
                "ACTA2", "COL1A1", "COL1A2", "FN1", "CDH1", "CDH5")
otherGenes_ENSG <- setNames(
  annotation_gene$id[annotation_gene$gene_name %in% otherGenes],
  annotation_gene$gene_name[annotation_gene$gene_name %in% otherGenes]
)

## SenMayo senescence markers group
senMayoGenes <- c("ACVR1B", "ANG", "ANGPT1", "ANGPTL4", "AREG", "AXL", "BEX3",
                  "BMP2", "BMP6", "C3", "CCL1", "CCL13", "CCL16",
                  "CCL20", "CCL24", "CCL26", "CCL3", "CCL3L1", "CCL4", "CCL5",
                  "CCL7", "CCL8", "CD55", "CD9", "CSF1", "CSF2", "CSF2RB",
                  "CST4", "CTNNB1", "CTSB", "CXCL1", "CXCL10", "CXCL12",
                  "CXCL16", "CXCL2", "CXCL3", "CXCL8", "CXCR2", "DKK1", "EDN1",
                  "EGF", "EGFR", "EREG", "ESM1", "ETS2", "FAS", "FGF1", "FGF2",
                  "FGF7", "GEM", "GMFG", "HGF", "HMGB1", "ICAM1",
                  "ICAM3", "IGF1", "IGFBP1", "IGFBP2", "IGFBP3", "IGFBP4",
                  "IGFBP5", "IGFBP6", "IGFBP7", "IL10", "IL13", "IL15", "IL18",
                  "IL1A", "IL1B", "IL2", "IL32", "IL7", "INHA",
                  "IQGAP2", "ITGA2", "ITPKA", "JUN", "KITLG", "LCP1", "MIF",
                  "MMP1", "MMP10", "MMP12", "MMP13", "MMP14", "MMP2", "MMP3",
                  "MMP9", "NAP1L4", "NRG1", "PAPPA", "PECAM1", "PGF", "PIGF",
                  "PLAT", "PLAU", "PLAUR", "PTBP1", "PTGER2", "PTGES",
                  "RPS6KA5", "SCAMP4", "SELPLG", "SEMA3F", "SERPINB4",
                  "SERPINE1", "SERPINE2", "SPP1", "SPX", "TIMP2", "TNF",
                  "TNFRSF10C", "TNFRSF11B", "TNFRSF1A", "TNFRSF1B", "TUBGCP2",
                  "VEGFA", "VEGFC", "VGF", "WNT16", "WNT2", "CCL2", "IL6",
                  "IL6ST", "GDF15")
senMayoGenes_ENSG <- setNames(
  annotation_gene$id[annotation_gene$gene_name %in% senMayoGenes],
  annotation_gene$gene_name[annotation_gene$gene_name %in% senMayoGenes]
)

## Epithelial to mesenchymal transition gene set from MSigDB HALLMARK GENESET
emtGenes <- c("ABI3BP", "ACTA2", "ADAM12", "ANPEP", "APLP1", "AREG", "BASP1",
              "BDNF", "BGN", "BMP1", "CADM1", "CALD1", "CALU", "CAP2", "CAPG",
              "CD44", "CD59", "CDH11", "CDH2", "CDH6", "COL11A1", "COL12A1",
              "COL16A1", "COL1A1", "COL1A2", "COL3A1", "COL4A1", "COL4A2",
              "COL5A1", "COL5A2", "COL5A3", "COL6A2", "COL6A3", "COL7A1",
              "COL8A2", "COMP", "COPA", "CRLF1", "CCN2", "CTHRC1", "CXCL1",
              "CXCL12", "CXCL6", "CCN1", "DAB2", "DCN", "DKK1", "DPYSL3",
              "DST", "ECM1", "ECM2", "EDIL3", "EFEMP2", "ELN", "EMP3", "ENO2",
              "FAP", "FAS", "FBLN1", "FBLN2", "FBLN5", "FBN1", "FBN2",
              "FERMT2", "FGF2", "FLNA", "FMOD", "FN1", "FOXC2", "FSTL1",
              "FSTL3", "FUCA1", "FZD8", "GADD45A", "GADD45B", "GAS1", "GEM",
              "GJA1", "GLIPR1", "COLGALT1", "GPC1", "GPX7", "GREM1", "HTRA1",
              "ID2", "IGFBP2", "IGFBP3", "IGFBP4", "IL15", "IL32", "IL6",
              "CXCL8", "INHBA", "ITGA2", "ITGA5", "ITGAV", "ITGB1", "ITGB3",
              "ITGB5", "JUN", "LAMA1", "LAMA2", "LAMA3", "LAMC1", "LAMC2",
              "P3H1", "LGALS1", "LOX", "LOXL1", "LOXL2", "LRP1", "LRRC15",
              "LUM", "MAGEE1", "MATN2", "MATN3", "MCM7", "MEST", "MFAP5",
              "MGP", "MMP1", "MMP14", "MMP2", "MMP3", "MSX1", "MXRA5", "MYL9",
              "MYLK", "NID2", "NNMT", "NOTCH2", "NT5E", "NTM", "OXTR", "PCOLCE",
              "PCOLCE2", "PDGFRB", "PDLIM4", "PFN2", "PLAUR", "PLOD1", "PLOD2",
              "PLOD3", "PMEPA1", "PMP22", "POSTN", "PPIB", "PRRX1", "PRSS2",
              "PTHLH", "PTX3", "PVR", "QSOX1", "RGS4", "RHOB", "SAT1", "SCG2",
              "SDC1", "SDC4", "SERPINE1", "SERPINE2", "SERPINH1", "SFRP1",
              "SFRP4", "SGCB", "SGCD", "SGCG", "SLC6A8", "SLIT2", "SLIT3",
              "SNAI2", "SNTB1", "SPARC", "SPOCK1", "SPP1", "TAGLN", "TFPI2",
              "TGFB1", "TGFBI", "TGFBR3", "TGM2", "THBS1", "THBS2", "THY1",
              "TIMP1", "TIMP3", "TNC", "TNFAIP3", "TNFRSF11B", "TNFRSF12A",
              "TPM1", "TPM2", "TPM4", "VCAM1", "VCAN", "VEGFA", "VEGFC", "VIM",
              "WIPF1", "WNT5A")
emtGenes_ENSG <- setNames(
  annotation_gene$id[annotation_gene$gene_name %in% emtGenes],
  annotation_gene$gene_name[annotation_gene$gene_name %in% emtGenes]
)


## GO Extracellular Matrix gene set from MSigDB GO::CC GENESET
ecmGenes <- c("A1BG", "A2M", "ABI3BP", "ACAN", "ACHE", "ADAM11", "ADAM19", "ADAMDEC1", "ADAMTS1",
              "ADAMTS10", "ADAMTS12", "ADAMTS13", "ADAMTS14", "ADAMTS15", "ADAMTS16", "ADAMTS17",
              "ADAMTS18", "ADAMTS19", "ADAMTS2", "ADAMTS20", "ADAMTS3", "ADAMTS4", "ADAMTS5", "ADAMTS6",
              "ADAMTS7", "ADAMTS8", "ADAMTS9", "ADAMTSL1", "ADAMTSL2", "ADAMTSL3", "ADAMTSL4", "ADAMTSL5",
              "ADIPOQ", "AEBP1", "AGRN", "AGT", "AHSG", "ALPL", "AMBP", "AMELX", "AMELY", "AMTN", "ANG",
              "ANGPT1", "ANGPT2", "ANGPT4", "ANGPTL1", "ANGPTL2", "ANGPTL3", "ANGPTL4", "ANGPTL5",
              "ANGPTL6", "ANGPTL7", "ANOS1", "ANXA1", "ANXA11", "ANXA2", "ANXA2P2", "ANXA4", "ANXA5",
              "ANXA6", "ANXA7", "ANXA8", "APCS", "APLP1", "APOA1", "APOA4", "APOC3", "APOE", "APOH",
              "ASPN", "ATRN", "ATRNL1", "AZGP1", "BCAM", "BCAN", "BGN", "BMP7", "BMPER", "C17orf58",
              "C1QA", "C1QB", "C1QC", "CALR", "CASK", "CBLN1", "CBLN4", "CCBE1", "CCDC80", "CCN1", "CCN2",
              "CCN3", "CCN4", "CCN5", "CCN6", "CD151", "CD180", "CD248", "CDH13", "CDH2", "CDON", "CFP",
              "CHAD", "CHADL", "CHI3L1", "CILP", "CLC", "CLEC14A", "CLEC3B", "CLU", "CMA1", "COCH",
              "COL10A1", "COL11A1", "COL11A2", "COL12A1", "COL13A1", "COL14A1", "COL15A1", "COL16A1",
              "COL17A1", "COL18A1", "COL19A1", "COL1A1", "COL1A2", "COL20A1", "COL21A1", "COL22A1",
              "COL23A1", "COL24A1", "COL25A1", "COL26A1", "COL27A1", "COL28A1", "COL2A1", "COL3A1",
              "COL4A1", "COL4A2", "COL4A3", "COL4A4", "COL4A5", "COL4A6", "COL5A1", "COL5A2", "COL5A3",
              "COL6A1", "COL6A2", "COL6A3", "COL6A5", "COL6A6", "COL7A1", "COL8A1", "COL8A2", "COL9A1",
              "COL9A2", "COL9A3", "COLEC12", "COLQ", "COMP", "CPA3", "CPN2", "CRELD1", "CRISP3",
              "CRISPLD2", "CSPG4", "CST3", "CSTB", "CTHRC1", "CTSB", "CTSC", "CTSD", "CTSF", "CTSG",
              "CTSH", "CTSL", "CTSS", "CTSZ", "CXCL12", "DAG1", "DCN", "DEFA1", "DEFA1B", "DGCR6",
              "DGCR6", "DLG1", "DMBT1", "DMP1", "DPT", "DSPP", "DST", "ECM1", "ECM2", "EDIL3", "EFEMP1",
              "EFEMP2", "EFNA5", "EGFL6", "EGFL7", "EGFLAM", "ELANE", "ELFN1", "ELFN2", "ELN", "EMID1",
              "EMILIN1", "EMILIN2", "EMILIN3", "ENAM", "ENTPD2", "EPYC", "ERBIN", "EYS", "F12", "F13A1",
              "F2", "F3", "F7", "F9", "FBLN1", "FBLN2", "FBLN5", "FBN1", "FBN2", "FBN3", "FCGBP", "FCN1",
              "FCN2", "FCN3", "FGA", "FGB", "FGF1", "FGF10", "FGF9", "FGFBP3", "FGFR2", "FGG", "FGL1",
              "FGL2", "FIBCD1", "FLG", "FLRT1", "FLRT2", "FLRT3", "FMOD", "FN1", "FRAS1", "FREM1",
              "FREM2", "FREM3", "GDF10", "GDF15", "GFOD2", "GH1", "GLG1", "GP1BA", "GPC1", "GPC2", "GPC3",
              "GPC4", "GPC5", "GPC6", "GPLD1", "GREM1", "HAPLN1", "HAPLN2", "HAPLN3", "HAPLN4", "HDGF",
              "HMCN1", "HMCN2", "HNRNPM", "HPSE", "HPSE2", "HPX", "HRG", "HRNR", "HSD17B12", "HSP90B1",
              "HSPG2", "HTRA1", "ICAM1", "IFNA2", "IGFALS", "IGFBP7", "IHH", "IL7", "IMPG1", "IMPG2",
              "INHBE", "ITGA6", "ITIH1", "ITIH2", "ITIH4", "ITIH5", "KAZALD1", "KERA", "KNG1", "KRT1",
              "L1CAM", "LAD1", "LAMA1", "LAMA2", "LAMA3", "LAMA4", "LAMA5", "LAMB1", "LAMB2", "LAMB3",
              "LAMB4", "LAMC1", "LAMC2", "LAMC3", "LEFTY2", "LGALS1", "LGALS3", "LGALS3BP", "LGALS4",
              "LGALS9", "LINGO1", "LINGO2", "LINGO3", "LINGO4", "LMAN1", "LMAN1L", "LOX", "LOXL1",
              "LOXL2", "LRIG1", "LRIG2", "LRIG3", "LRRC15", "LRRC17", "LRRC24", "LRRC32", "LRRC3B",
              "LRRC3C", "LRRN1", "LRRN2", "LRRN3", "LRRTM1", "LRRTM3", "LRRTM4", "LTBP1", "LTBP2",
              "LTBP3", "LTBP4", "LUM", "MARCOL", "MATN1", "MATN2", "MATN3", "MATN4", "MBL2", "MDK",
              "MEGF9", "MEPE", "MFAP1", "MFAP2", "MFAP4", "MFAP5", "MFGE8", "MGP", "MMP1", "MMP10",
              "MMP11", "MMP12", "MMP13", "MMP14", "MMP15", "MMP16", "MMP17", "MMP19", "MMP2", "MMP20",
              "MMP21", "MMP23B", "MMP24", "MMP25", "MMP26", "MMP27", "MMP28", "MMP3", "MMP7", "MMP8",
              "MMP9", "MMRN1", "MMRN2", "MST1", "MUC17", "MUC2", "MUC4", "MUC5AC", "MUC5B", "MUC6",
              "MXRA5", "MXRA7", "MYOC", "NAV2", "NCAM1", "NCAN", "NDNF", "NDP", "NID1", "NID2", "NPNT",
              "NPPA", "NTN1", "NTN3", "NTN4", "NTN5", "NTNG1", "NTNG2", "NYX", "OC90", "OGN", "OLFML2A",
              "OMD", "OPTC", "ORM1", "ORM2", "OTOG", "OTOGL", "OTOL1", "P3H1", "P3H2", "PAPLN", "PCOLCE",
              "PCSK6", "PDGFB", "PF4", "PHOSPHO1", "PI3", "PKM", "PLG", "PLOD3", "PLSCR1", "PODN",
              "PODNL1", "POMZP3", "POSTN", "PRELP", "PRG2", "PRG3", "PRG4", "PRSS1", "PRSS2", "PRTN3",
              "PSAP", "PTN", "PTPRZ1", "PTX3", "PXDN", "PZP", "RARRES2", "RBP3", "RELL2", "RELN", "RTBDN",
              "RTN4RL1", "RTN4RL2", "S100A10", "S100A4", "S100A6", "S100A7", "S100A8", "S100A9", "SBSPON",
              "SCARA3", "SDC2", "SDC3", "SEMA3B", "SEMA7A", "SERAC1", "SERPINA1", "SERPINA3", "SERPINA5",
              "SERPINB1", "SERPINB12", "SERPINB6", "SERPINB8", "SERPINB9", "SERPINC1", "SERPINE1",
              "SERPINE2", "SERPINF1", "SERPINF2", "SERPING1", "SERPINH1", "SFRP1", "SFRP2", "SHH", "SLPI",
              "SMOC1", "SMOC2", "SNORC", "SOD3", "SOST", "SPARC", "SPARCL1", "SPOCK2", "SPON1", "SPON2",
              "SPP2", "SRPX", "SRPX2", "SSC5D", "SSPOP", "SULF1", "TECTA", "TECTB", "TFIP11", "TFPI2",
              "TGFB1", "TGFB1I1", "TGFB2", "TGFB3", "TGFBI", "TGFBR3", "TGM2", "TGM4", "THBS1", "THBS2",
              "THBS3", "THBS4", "THSD4", "TIMP1", "TIMP2", "TIMP3", "TIMP4", "TINAG", "TINAGL1", "TLR3",
              "TMEFF1", "TMEFF2", "TNC", "TNFRSF11B", "TNN", "TNR", "TNXB", "TPSAB1", "TPSB2", "TRIL",
              "UCMA", "USH2A", "VASN", "VCAN", "VEGFA", "VIT", "VTN", "VWA1", "VWA2", "VWC2", "VWF",
              "WNT11", "WNT2", "WNT2B", "WNT3", "WNT4", "WNT5A", "WNT5B", "WNT6", "WNT7A", "WNT8A", "ZAN",
              "ZG16", "ZP1", "ZP2", "ZP3")

ecmGenes_ENSG <- setNames(
  annotation_gene$id[annotation_gene$gene_name %in% ecmGenes],
  annotation_gene$gene_name[annotation_gene$gene_name %in% ecmGenes]
) # Removed "ANXA2P2" "SSPOP" because are marked as Pseudogenes


## TGF-beta & IL4 gene set as found in : DOI:
tgfbGenes <- c("AC003092.1", "AC022509.2", "AC083837.2", "ADAMTS14", "ADAMTS4", "ADAMTS9", "AMPD3",
               "ANGPTL4", "ANO1", "APOL1", "ATF3", "BATF2", "BCL2A1", "BDKRB1", "BHLHE40-AS1", "BIRC3",
               "C15orf48", "C3", "C5orf56", "CACNA1G", "CARD6", "CAVIN4", "CCL11", "CCL2", "CCL5", "CCL7",
               "CD274", "CD38", "CD70", "CD82", "CD83", "CFAP45", "CHI3L2", "CLDN1", "CLDN4", "CMPK2",
               "COL22A1", "CSF1", "CSF2", "CTSS", "CX3CL1", "CXCL1", "CXCL10", "CXCL2", "CXCL3", "CXCL6",
               "CXCL8", "CYP7B1", "DCLK3", "DDX58", "DDX60", "DDX60L", "EDARADD", "EGR2", "ENC1", "EREG",
               "ESM1", "FAM167A", "G0S2", "GBP1P1", "GFPT2", "GPR68", "GRIP2", "HAS1", "HBEGF", "HCK",
               "HELZ2", "HLA-F", "ICAM1", "IER3", "IFIH1", "IFIT2", "IFIT3", "IL11", "IL15", "IL15RA",
               "IL18BP", "IL1B", "IL32", "IL33", "IL6", "IRAK2", "IRF7", "ITGB3", "JUNB", "KCTD11", "KDR",
               "L1TD1", "LGALS9", "LIF", "LINC01539", "LIPG", "LRIG1", "MCC", "MED12L", "MEOX1", "MFSD2A",
               "MMP12", "MSC", "MX1", "MX2", "MYO10", "NFKB2", "NFKBIA", "NFKBIE", "NFKBIZ", "NGFR",
               "NIPAL4", "NKD1", "NKX3-1", "NPAS2", "NPTX1", "NR4A1", "NR4A2", "OAS1", "OASL", "OGFR",
               "P2RY6", "PARP12", "PARP14", "PDPN", "PHACTR1", "PKP1", "POU2F2", "PREX1", "PTGS2", "PTHLH",
               "RASD1", "RASL11B", "RCSD1", "RELB", "RFLNA", "RGS16", "RIPK2", "RIPOR3", "RND1", "RNF19B",
               "RRAD", "RSPO3", "SAA1", "SERPINA1", "SERPINB2", "SLAMF8", "SLC12A7", "SLC22A3", "SLC25A28",
               "SLC2A6", "SLC39A14", "SMOX", "SOD2", "SP6", "STAT5A", "STRA6", "STX11", "SYT12", "TAP1",
               "TCIM", "TFPI2", "TGFA", "TLR3", "TMEM51", "TNF", "TNFAIP2", "TNFAIP3", "TNFAIP6",
               "TNFAIP8", "TNIP1", "TRAF1", "TRIM36", "TRPA1", "VCAM1", "XAF1", "ZC3H12A", "ZFPM2", "ZNF710")

tgfbGenes_ENSG <- setNames(
  annotation_gene$id[annotation_gene$gene_name %in% tgfbGenes],
  annotation_gene$gene_name[annotation_gene$gene_name %in% tgfbGenes]
) # Removed "AC083837.2" "GBP1P1" because are marked as Pseudogenes

## Senescence-related transcription factors
transcrFact <- c("SNAI1", "SNAI2", "TWIST1", "ZEB1", "ZEB2")
transcrFact_ENSG <- setNames(
  annotation_gene$id[annotation_gene$gene_name %in% transcrFact],
  annotation_gene$gene_name[annotation_gene$gene_name %in% transcrFact]
)

## Concatenate all but the ligand-receptor vector
selectedGenes <- c(senescenceMarkers, otherGenes, senMayoGenes, transcrFact)
selectedGenes_ENSG <- c(senescenceMarkers_ENSG, otherGenes_ENSG,
                        senMayoGenes_ENSG, transcrFact_ENSG)

# ---------------------------------------------------------------------------- #
# ----Select cell types --------------------------------------------------------
cellTypes <- colnames(cellAbundance[[1]][3:ncol(cellAbundance[[1]])]) %>%
  gsub("q05cell_abundance_w_sf_", "", .)
names(cellTypes) <- cellGroups$cell_name_special

selectedCellTypes <- c("B.Cells", "Basal", "Fibroblasts", "Endothelial.Cells",
                       "Macrophages", "T.Cells", "AT1", "AT2", "MUC5B.",
                       "KRT5..KRT17.", "Transitional.AT2", "Myofibroblasts",
                       "Ciliated", "HAS1.High.Fibroblasts",
                       "MUC5AC..High", "SCGB3A2..SCGB1A1.", "SCGB3A2.",
                       "Proliferating.Epithelial.Cells")
names(selectedCellTypes) <- c("B cells", "Basal", "Fibroblasts",
                              "Endothelial cells", "Macrophages", "T cells",
                              "AT1", "AT2", "MUC5B+", "KRT5-/KRT17+",
                              "Transitional AT2", "Myofibroblasts",
                              "Ciliated", "HAS1 High Fibroblasts",
                              "MUC5AC-high", "SCGB3A2+SCGB1A1+", "SCGB3A2+",
                              "Proliferating Epithelial Cells")

# ---------------------------------------------------------------------------- #
# ----Custom functions to use --------------------------------------------------
# Define a function to print caught errors
error = function(e){cat("ERROR: ", conditionMessage(e), "\n")}

# Define a function to calculate the proportion of spots expressing each gene
calculate_presence_proportion <- function(sfe, genes, threshold = 0) {
  # Subset the dataset to include only the genes of interest
  gene_expr <- as.matrix(assay(sfe, "logcounts"))[genes, ]

  # Determine the spots where each gene is expressed above the threshold
  expressed <- gene_expr > threshold

  # Calculate the proportion of spots where each gene is expressed
  presence_proportion <- rowMeans(expressed)

  return(presence_proportion)
}

# Define a function to convert from Seurat to SFE.
# It is quick and dirty. For some reason the toSpatialFeatureExperiment from the
# SpatialFeatureExperiment package returns NULL.
toSFE <- function(seu, ...) {
  ## Convert to SCE object
  sc <- Seurat::as.SingleCellExperiment(seu)

  sfe <- SpatialFeatureExperiment::toSpatialFeatureExperiment(sc,
                                    ...)
  return(sfe)
}


# ---------------------------------------------------------------------------- #
# ----Analysis -----------------------------------------------------------------


# ---------------------------------------------------------------------------- #
## ----QC ----------------------------------------------------------------------
### ----Calculate QC Metrics ---------------------------------------------------
## Mark a subset of mitochondrial genes
is_mito <- getSubset(msfe,
                     sample_id = TRUE,
                     subset = "(^MT-)|(^mt-)|(^MTRNR)",
                     set = "rowData",
                     col_name = "symbol")
## Mark a subset of haemoglobin genes
is_hemo <- getSubset(msfe,
                     sample_id = TRUE,
                     subset = "^HB",
                     set = "rowData",
                     col_name = "symbol")
## Mark a subset of Ribosomal genes
is_ribo <- getSubset(msfe,
                     sample_id = TRUE,
                     subset = "^RPS|^RPL",
                     set = "rowData",
                     col_name = "symbol")

for (id in sampleNames_selected) {
  message("Working on sample: ", id)
  ## Add location-related statistics
  msfe <- addPerLocQC(msfe,
                     sample_id = id,
                     gTruth = gTruth_list[[id]],
                     assay = "counts",
                     MARGIN = 2,
                     subsets = list(mito = is_mito[[id]],
                                    hemo = is_hemo[[id]],
                                    ribo = is_ribo[[id]]))
  message("\tAdded location-related statistics")

  ## Add geometries
  msfe <- addGeometries(msfe,
                       samples = sampleDir[id],
                       sample_id = id,
                       res = "fullres",
                       flipped = FALSE,
                       geoms = "both")
  message("\tAdded geometries")

  ## Add gene/feature-related statistics
  msfe <- addPerGeneQC(msfe,
                      sample_id = id,
                      assay = "counts",
                      version = NULL,
                      mirror = NULL,
                      add = "none")
  message("\tAdded gene/feature-related statistics")
}

## Check that annotations were successfully added
for (i in seq_along(sampleNames_selected)) {
  message(sampleNames_selected[i], ": ",
          "annotation" %in% colnames(colData(msfe@sfe_data[[i]])))
}


# ---------------------------------------------------------------------------- #
### ----Quality Control: Spots--------------------------------------------------
#### ----Library size ----------------------------------------------------------
## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/QC_Spots/QC_library_size/")
dir.create(graphics_out, recursive = TRUE)

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Extract sample
  sfe <- getSFE(msfe, sample_id = id)

  ## Select filter
  min <- 350 # at least 350 unique molecules per spot

  ## Density and histogram of library sizes
  p1 <- plotQC_hist(sfe, metric = "libsize",
                    limits = c(min, NA),
                    hist_args = list(bins = 100)) +
    theme(axis.text.x = element_text(size = 12, angle = 45))

  ## Map the library sizes
  p2 <- plotQC_map(sfe, metric = "libsize", type = "hex", size = 0.5)

  ## Select threshold
  sfe <- setQCthresh_LibSize(sfe, sample_id = TRUE,
                             min_t = min,
                             max_t = NA)

  ## Check putative spatial patterns of removed spots
  p3 <- plotQC_filtered(sfe, metric = "libsize", sample_id = TRUE, type = "hex", size = 0.5)

  ## Plot tissue image
  p4 <- plotQC_tissueImg(sfe, res = "lowres", sample_id = id, type = "none")

  ## Patch, print and save
  print((p1 + p2) / (p3 + p4))

  prfx <- id
  main <- "_QC_LibSize"
  sfx <- ""
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = grDevices::dev.size(units = "in")[1],
         height = grDevices::dev.size(units = "in")[2],
         units = "in",
         dpi = 300)

  ## Update msfe object
  msfe@sfe_data[[id]] <- sfe

  ## Housekeeping
  rm(sfe, p1, p2, p3, p4)
}


# ---------------------------------------------------------------------------- #
#### ----Number of expressed genes ---------------------------------------------
## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/QC_Spots/QC_gene_number/")
dir.create(graphics_out, recursive = TRUE)

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Extract sample
  sfe <- getSFE(msfe, sample_id = id)

  ## Select filter
  min <- 100 # at least 100 genes per spot

  ## Density and histogram of expressed genes
  p1 <- plotQC_hist(sfe, metric = "detected", c(min, NA))

  ## Map the library sizes
  p2 <- plotQC_map(sfe, metric = "detected", type = "hex", size = 0.5)

  ## Select threshold
  sfe <- setQCthresh_GenesExpr(sfe, sample_id = TRUE,
                               min_t = min,
                               max_t = NA)

  ## Check putative spatial patterns of removed spots
  p3 <- plotQC_filtered(sfe, metric = "detected", sample_id = TRUE, type = "hex", size = 0.5)

  ## Patch, print and save
  print(p1 + p2 + p3)

  prfx <- id
  main <- "_QC_GeneNumber"
  sfx <- ""
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = grDevices::dev.size(units = "in")[1],
         height = grDevices::dev.size(units = "in")[2],
         units = "in",
         dpi = 300)

  ## Update msfe object
  msfe@sfe_data[[id]] <- sfe

  ## Housekeeping
  rm(sfe, p1, p2, p3)
}


# ---------------------------------------------------------------------------- #
#### ----Percent of mito expression --------------------------------------------
## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/QC_Spots/QC_mito_percent/")
dir.create(graphics_out, recursive = TRUE)

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Extract sample
  sfe <- getSFE(msfe, sample_id = id)

  ## Density and histogram of percentage of mitochondrial expression
  p1 <- plotQC_hist(sfe, metric = "mito", limits = c(NA, 15))

  ## Map the library sizes
  p2 <- plotQC_map(sfe, metric = "mito", type = "hex", size = 0.5)

  ## Select threshold
  sfe <- setQCthresh_Mito(sfe, sample_id = TRUE, min_t = NA, max_t = 15)

  ## Check putative spatial patterns of removed spots
  p3 <- plotQC_filtered(sfe, metric = "mito", sample_id = TRUE, type = "hex", size = 0.5)

  ## Patch, print and save
  print(p1 + p2 + p3)

  prfx <- id
  main <- "_QC_MitoPercent"
  sfx <- ""
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = grDevices::dev.size(units = "in")[1],
         height = grDevices::dev.size(units = "in")[2],
         units = "in",
         dpi = 300)

  ## Update msfe object
  msfe@sfe_data[[id]] <- sfe

  ## Housekeeping
  rm(sfe, p1, p2, p3)
}


# ---------------------------------------------------------------------------- #
#### ----Percent of haemoglobin expression -------------------------------------
## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/QC_Spots/QC_hemo_percent/")
dir.create(graphics_out, recursive = TRUE)

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Extract sample
  sfe <- getSFE(msfe, sample_id = id)

  ## Get the vector
  hemo_out <- list(id = sfe@colData$subsets_hemo_percent > 30)
  names(hemo_out) <- id

  ## Map the library sizes
  p1 <- plotQC_map(sfe, metric = "custom", type = "hex", size = 0.5, metric_name = "subsets_hemo_percent")

  ## Select threshold
  sfe <- setQCthresh_custom(sfe, sample_id = TRUE, MARGIN = 2, qcMetric = hemo_out)

  ## Check putative spatial patterns of removed spots
  p2 <- plotQC_filtered(sfe,
                        metric = "custom",
                        sample_id = TRUE,
                        type = "hex",
                        size = 0.5,
                        metric_name = "qc_hemo_out",
                        metric_lab = "Filtered for % Hemoglobin Expression")

  ## Patch, print and save
  print(p1 + p2)

  prfx <- id
  main <- "_QC_HemoPercent"
  sfx <- ""
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = grDevices::dev.size(units = "in")[1],
         height = grDevices::dev.size(units = "in")[2],
         units = "in",
         dpi = 300)

  ## Update msfe object
  msfe@sfe_data[[id]] <- sfe

  ## Housekeeping
  rm(sfe, p1, p2, hemo_out)
}


# ---------------------------------------------------------------------------- #
#### ----Remove low-quality spots ----------------------------------------------
## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/QC_Spots/QC_filter_spots/")
dir.create(graphics_out, recursive = TRUE)

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Extract sample
  sfe <- getSFE(msfe, sample_id = id)

  ## Set the combined filtering threshold using the QC metrics
  sfe <- setQCtoDiscard_loc(sfe, sample_id = TRUE, filters = TRUE)

  ## Check putative spatial patterns of removed spots
  plotQC_filtered(sfe, metric = "discard", sample_id = TRUE, type = "hex")

  prfx <- id
  main <- "_QC_SpotsFiltered"
  sfx <- ""
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = grDevices::dev.size(units = "in")[1],
         height = grDevices::dev.size(units = "in")[2],
         units = "in",
         dpi = 300)

  ## Remove combined set of low-quality spots
  sfe <- applyQCthresh_loc(sfe, sample_id = TRUE)

  ## Update msfe object
  msfe@sfe_data[[id]] <- sfe

  ## Housekeeping
  rm(sfe)
}


# ---------------------------------------------------------------------------- #
### ----Quality Control: Features ----------------------------------------------

# ---------------------------------------------------------------------------- #
#### ----Mark specific gene sets -----------------------------------------------
## Genes in the XY chromosomes
xychr_genes <- unique(subset(annotation_gene, annotation_gene$chromosome %in% c("X", "Y"))$gene_name)
is_xy <- lapply(sampleNames_selected, function(x){rowData(msfe@sfe_data[[x]])[["gene_name"]] %in% xychr_genes})
names(is_xy) <- sampleNames_selected

## Keep only protein coding, IG and TR genes
keep_biotype <- c("protein_coding", grep("_gene", unique(annotation_gene$biotype), value = TRUE))
coding_genes <- annotation_gene[annotation_gene$biotype %in% keep_biotype, "gene_name"]
is_ncGene <- lapply(sampleNames_selected, function(x){!rowData(msfe@sfe_data[[x]])[["gene_name"]] %in% coding_genes$gene_name})
names(is_ncGene) <- sampleNames_selected

## Remove mitochondrial, hemoglobin, ribosomal, XY, and anything NOT protein coding or IG or TR.
for (id in sampleNames_selected) {
  msfe <- setQCthresh_custom(msfe, MARGIN = 1, qcMetric = is_mito[[id]])
  msfe <- setQCthresh_custom(msfe, MARGIN = 1, qcMetric = is_hemo[[id]])
  msfe <- setQCthresh_custom(msfe, MARGIN = 1, qcMetric = is_ribo[[id]])
  msfe <- setQCthresh_custom(msfe, MARGIN = 1, qcMetric = is_xy[[id]])
  msfe <- setQCthresh_custom(msfe, MARGIN = 1, qcMetric = is_ncGene[[id]])
}

# ---------------------------------------------------------------------------- #
#### ----Remove low-quality genes ----------------------------------------------
## QC discard Features
## Set the combined filtering threshold using the QC metrics
msfe <- setQCtoDiscard_feat(msfe, filters = TRUE)

## Apply gene-level QC threshold
msfe <- applyQCthresh_feat(msfe)


# ---------------------------------------------------------------------------- #
## ----Normalisation of Counts -------------------------------------------------
### ----Get MultiSFE object ----------------------------------------------------
sfe_multi <- getMultiSFE(msfe, sampleNames_selected)

## Add the subject alias in the colData to regress it out downstream
sfe_multi@colData$subject_alias <- left_join(as.data.frame(sfe_multi@colData["sample_id"]),
                                             metadata,
                                             by = "sample_id")[["subject_alias"]]

## Remove genes with zero expression in all samples
sfe_multi <- setQCthresh_ZeroExpr(sfe_multi)
sfe_multi <- setQCtoDiscard_feat(sfe_multi)
sfe_multi <- applyQCthresh_feat(sfe_multi)

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(sfe_multi, "../STExplorer_Analysis/Lung/analysis_objs/sfe_multi.rds")

### ----Convert to Seurat object -----------------------------------------------
seu <- as.Seurat(sfe_multi, data = NULL)


### ----SCTransform counts -----------------------------------------------------
## Normalise and regress out sampleID and donor, to remove major effects from
## technical and interindividual differences.
vars_to_reg <- c("sample_id", "subject_alias")

options(future.globals.maxSize = 8000 * 1024^2)

seu <- SCTransform(seu, vars.to.regress = vars_to_reg, assay = "originalexp", conserve.memory = TRUE)


## ----Generate Module Scores --------------------------------------------------
### ----Senescence -------------------------------------------------------------
seu <- AddModuleScore(seu, features = list(senMayo = senMayoGenes_ENSG), assay = "SCT", name = "senMayo")

### ----EMT --------------------------------------------------------------------
seu <- AddModuleScore(seu, features = list(emt = emtGenes_ENSG), assay = "SCT", name = "emt")

### ----ECM --------------------------------------------------------------------
seu <- AddModuleScore(seu, features = list(ecm = ecmGenes_ENSG), assay = "SCT", name = "ecm")

### ----ECM --------------------------------------------------------------------
seu <- AddModuleScore(seu, features = list(tgfb = tgfbGenes_ENSG), assay = "SCT", name = "tgfb")

## Some genes from the above feature sets are not present in the data set.
## These genes have been filtered out. However, it is good to know which are
## these genes:
cat("Genes from SenMayo gene set not present in the dataset:", sort(names(senMayoGenes_ENSG[!senMayoGenes_ENSG %in% rownames(seu)])), "\n")
cat("Genes from EMT gene set not present in the dataset:", sort(names(emtGenes_ENSG[!emtGenes_ENSG %in% rownames(seu)])), "\n")
cat("Genes from ECM gene set not present in the dataset:", sort(names(ecmGenes_ENSG[!ecmGenes_ENSG %in% rownames(seu)])), "\n")
cat("Genes from TGFB gene set not present in the dataset:", sort(names(tgfbGenes_ENSG[!tgfbGenes_ENSG %in% rownames(seu)])), "\n")

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(seu, file = "../STExplorer_Analysis/Lung/analysis_objs/seu.rds")

## ----Convert back to SFE -----------------------------------------------------
## Move the SCT counts and the senMayo/emt/ecm/tgfb scores into the sfe_multi object.
## Then split the sfe_multi object into multiple SFE objects inside a MetaSFE

## 1. First remove the genes that have been filtered out after SCTransform
to_keep <- rownames(seu)
sfe_multi <- sfe_multi[to_keep,]

## 2. Add the SCTransformed counts
assay(sfe_multi, "sct") <- seu[["SCT"]]$data

## 3. Add the senMayo/emt/ecm/tgfb
colData(sfe_multi)$senMayo <- as.vector(seu$senMayo1)
colData(sfe_multi)$emt <- as.vector(seu$emt1)
colData(sfe_multi)$ecm <- as.vector(seu$ecm1)
colData(sfe_multi)$tgfb <- as.vector(seu$tgfb1)

## 4. Split into single SFEs inside an MSFE object
msfe <- addMultiSFE(msfe, sfe_multi)

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(sfe_multi, "../STExplorer_Analysis/Lung/analysis_objs/sfe_multi.rds")
saveRDS(msfe, "../STExplorer_Analysis/Lung/analysis_objs/msfe.rds")

# ---------------------------------------------------------------------------- #
## ----Add Distance Matrices ---------------------------------------------------
## Calculate a simple distance matrix
msfe <- addDistMat(msfe, p = 2)

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(msfe, file = paste0(projectFolder, "msfe_QCed.rds"))

# ---------------------------------------------------------------------------- #
## ----Spatial Autocorrelation -------------------------------------------------
### ----Global Moran's I -------------------------------------------------------
sa_moran_gs <- list()

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Calculate Moran's I
  msfe <- moranGlobalIPerm(msfe, sample_id = id, genes = selectedGenes_ENSG, mc.cores = 7)

  ## Retrieve genes with significant positive SA
  sa_moran_gs[[id]] <- getSAGlobalGenes(m_sfe = msfe,
                                        sample_id = id,
                                        statistic = "moran",
                                        test = "permutation",
                                        pVal = 0.05,
                                        stat_thresh = 0.3)
}


### ----Global Geary's C -------------------------------------------------------
sa_geary_gs <- list()

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Calculate Geary's C
  msfe <- gearyGlobalCPerm(msfe, sample_id = id, genes = selectedGenes_ENSG, mc.cores = 7)

  ## Retrieve genes with significant positive SA
  sa_geary_gs[[id]] <- getSAGlobalGenes(m_sfe = msfe,
                                        sample_id = id,
                                        statistic = "geary",
                                        test = "permutation",
                                        pVal = 0.05,
                                        stat_thresh = NULL)
}


### ----Global Getis & Ord's G -------------------------------------------------
sa_getis_gs <- list()

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  msfe <- getisGlobalGTest(msfe, sample_id = id, genes = selectedGenes_ENSG, mc.cores = 7)

  sa_getis_gs[[id]] <- getSAGlobalGenes(m_sfe = msfe,
                                        sample_id = id,
                                        statistic = "getis",
                                        test = "z-score",
                                        pVal = 0.05,
                                        stat_thresh = NULL)
}


### ----Global Stats Summary ---------------------------------------------------
# statistic <- c("moran", "getis", "geary")
statistic <- "getis"

## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/SA/SA_Global_Summary/")
dir.create(graphics_out, recursive = TRUE)

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ### Get genes to be used --> combination of all global statistics
  if ("moran" %in% statistic) {
    gs_moran <- sa_moran_gs[[id]]
  } else {
    gs_moran <- NULL
  }
  if ("geary" %in% statistic) {
    gs_geary <- sa_geary_gs[[id]]
  } else {
    gs_geary <- NULL
  }
  if ("getis" %in% statistic) {
    gs_getis <- sa_getis_gs[[id]]
  } else {
    gs_getis <- NULL
  }

  ## A single vector with all genes from all SA statistics
  gs_to_plot <- unique(c(gs_moran, gs_geary, gs_getis))

  plotSA_globalSum(msfe,
                   sample_id = id,
                   genes = gs_to_plot,
                   statistic = statistic,
                   ylab = "gene_name",
                   moran_thresh = c(-0.5, 0.5),
                   geary_thresh = c(0.5, 1.5),
                   getis_thresh = 2)

  prfx <- id
  main <- "_SA_GlobalSum"
  sfx <- paste0("_", paste0(statistic, collapse = "_"))
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = grDevices::dev.size(units = "in")[1],
         height = grDevices::dev.size(units = "in")[2],
         units = "in",
         dpi = 300)
}
## Housekeeping
rm(gs_to_plot, gs_moran, gs_geary, gs_getis)


### ----Local Moran's I --------------------------------------------------------
## Run Local statistic only for the genes that where selected from the global
## statistic
for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  msfe <- moranLocalIPerm(m_sfe = msfe,
                          sample_id = id,
                          genes = selectedGenes_ENSG,
                          mc.cores = 6)
}


### ----Local Geary's C --------------------------------------------------------
## Run Local statistic only for the genes that where selected from the global
## statistic
for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  msfe <- gearyLocalCPerm(m_sfe = msfe,
                          sample_id = id,
                          genes = selectedGenes_ENSG,
                          mc.cores = 6)
}


### ----Local Getis & Ord's G --------------------------------------------------
## Run Local statistic only for the genes that where selected from the global
## statistic
for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  msfe <- getisLocalGPerm(m_sfe = msfe,
                          sample_id = id,
                          genes = selectedGenes_ENSG,
                          mc.cores = 6)
}

### ----Plot SA ----------------------------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  for (t in c("moran", "geary", "getis")) {
    if (t == "moran") {
      message("\t... on Moran's I")
      localRes <- "localMoranIPerm"
      graphics_out <- paste0(projectFolder, "/graphics_out/SA/SA_Local_MoranI/")
      main <- "_SA_LocalI"

    } else if (t == "geary") {
      message("\t... on Geary's C")
      localRes <- "localGearyCPerm"
      graphics_out <- paste0(projectFolder, "/graphics_out/SA/SA_Local_GearyC/")
      main <- "_SA_LocalC"

    } else if (t == "getis") {
      message("\t... on Getis & Ord's G")
      localRes <- "localGetisGPerm"
      graphics_out <- paste0(projectFolder, "/graphics_out/SA/SA_Local_GetisG/")
      main <- "_SA_LocalG"
    }

    for (g in selectedGenes_ENSG) {
      message(paste0("\t\t... on gene: ", g))
      g_name <- names(grep(g, selectedGenes_ENSG, value = TRUE))
      out_path <- paste0(graphics_out, g_name, "/")
      dir.create(out_path, recursive = TRUE)
      ## Check gene is present in the results
      isPresent <- g %in% names(localResults(msfe@sfe_data[[id]])[[localRes]])
      if (isPresent) {
        ## Plot
        for (l in c("all", "significant")) {
          p1 <- plotSA_local(m_sfe = msfe,
                             sample_id = id,
                             feature = g,
                             statistic = t,
                             test = "permutation",
                             pVal = 0.05,
                             type = "hex",
                             title = "name",
                             locations = l)

          p2 <- plotSA_localClust(m_sfe = msfe,
                                  sample_id = id,
                                  feature = g,
                                  statistic = t,
                                  test = "permutation",
                                  pVal = 0.05,
                                  type = "hex",
                                  title = "name",
                                  clust_col = NULL,
                                  locations = l)

          print((p1 / p2))

          prfx <- id
          main <- main
          sfx <- paste0("_", g_name)
          other <- ifelse(l == "all", "_all", "_sig")

          ggsave(paste0(out_path, prfx, main, sfx, other, ".png"),
                 device = "png",
                 width = 5.22,
                 height = 8.23,
                 units = "in",
                 dpi = 300)
        }
      } else {
        next
      }
    }
  }
}
## Housekeeping
rm( p1, p2)

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(msfe, file = paste0(projectFolder, "msfe_SA.rds"))


# ---------------------------------------------------------------------------- #
## ----GWR ---------------------------------------------------------------------
### ----Gene VS Gene -----------------------------------------------------------
#### ----Select genes ----------------------------------------------------------
selectedGenes_GWR <- tidyr::expand_grid(senescenceMarkers_ENSG, otherGenes_ENSG)

#### ----Build formulas --------------------------------------------------------
formulas_genes <- paste(selectedGenes_GWR$senescenceMarkers_ENSG,
                  selectedGenes_GWR$otherGenes_ENSG,
                  sep = "~")
names <- paste(names(selectedGenes_GWR$senescenceMarkers_ENSG),
               names(selectedGenes_GWR$otherGenes_ENSG),
               sep = "~")
names(formulas_genes) <- names

## Add a few more pairs of genes
formulas_genes2 <- c("ENSG00000039068~ENSG00000108821", "ENSG00000039068~ENSG00000164692", "ENSG00000039068~ENSG00000107796",
                    "ENSG00000179776~ENSG00000108821", "ENSG00000179776~ENSG00000164692", "ENSG00000179776~ENSG00000107796",
                    "ENSG00000107796~ENSG00000136244", "ENSG00000107796~ENSG00000108691", "ENSG00000107796~ENSG00000105329")
names(formulas_genes2) <- c("CDH1~COL1A1", "CDH1~COL1A2", "CDH1~ACTA2",
                           "CDH5~COL1A1", "CDH5~COL1A2", "CDH5~ACTA2",
                           "ACTA2~IL6", "ACTA2~CCL2", "ACTA2~TGFB1")

## Append them to the original list
formulas_genes <- c(formulas_genes, formulas_genes2)

#### ----Run GWR ---------------------------------------------------------------
gwr_list_genes <- vector(mode = "list", length = length(sampleNames_selected))
gwr_runs <- vector(mode = "list", length = length(formulas_genes))
names(gwr_list_genes) <- sampleNames_selected
names(gwr_runs) <- names(formulas_genes)

for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  for (i in seq_along(formulas_genes)) {
    ## Some formulas return `Error: inv(): matrix is singular`. This is because
    ## there are too few values in the area to generate a regression model.
    ## For the sake of running everything once and tackling the erring formulas
    ## one by one, we wrap everything inside a tryCatch function.
    tryCatch({
      message("\tWorking on formula: ", names(formulas_genes[i]), " (", i, "/", length(formulas_genes), ")")

      ### ----Set Bandwidth ---------------------------------------------------#
      message("\t\tSetting bandwidth...")
      spot_diameter <- msfe@sfe_data[[id]]@metadata[["spotDiameter"]][[id]][["spot_diameter_fullres"]]
      bw <- 3*spot_diameter


      ### ----Run GWR ---------------------------------------------------------#
      message("\t\tRunning GWR...")
      gwr_runs[[names(formulas_genes[i])]] <- gwrSTE(gwr_method = "basic",
                                                     formula = formulas_genes[[i]],
                                                     m_sfe = msfe,
                                                     sample_id = id,
                                                     bw = bw,
                                                     kernel = "exponential")
    }, error = error)
  }

  ## Update output list
  message("Updating the list of GWR runs per sample.")
  gwr_list_genes[[id]] <- gwr_runs
}

## Housekeeping
rm(spot_diameter, gwr_runs2)

#### ----Plot GWR --------------------------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Select GWR run for the specific sample
  gwrRuns <- gwr_list_genes[[id]]

  for (i in seq_along(formulas_genes)) {
    message("\tWorking on formula: ", names(formulas_genes[i]), " (", i, "/", length(formulas_genes), ")")

    ## Create folder
    graphics_out <- paste0(projectFolder, "/graphics_out/GWR/", names(formulas_genes[i]), "/")
    dir.create(graphics_out, recursive = TRUE)

    ## Plot GWR coefficients
    plotGWR(gwrRuns[[i]])

    prfx <- id
    main <- "_GWR"
    sfx <- paste0("_", names(formulas_genes[i]))
    other <- ""

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = grDevices::dev.size(units = "in")[1],
           height = grDevices::dev.size(units = "in")[2],
           units = "in",
           dpi = 300)
  }
  ## Housekeeping
  rm(gwRuns)
}


# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(gwr_list_genes, file = paste0(projectFolder, "/gwr_list_genes.rds"))


### ----Cell VS Cell (abundance) -----------------------------------------------
#### ----Import cell abundances ------------------------------------------------
## First we need to import the cell abundances into the colData
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Extract abundances for specific cell types -------------------------------#
  message("\tExtracting abundances...")
  ### Extract barcodes from colData because SFEs are QC'ed but cell abundances not
  spotBarcodes <- colData(msfe@sfe_data[[id]])[["Barcode"]]
  ### Extract abundances
  abundance_df <- cellAbundance[[id]] %>%
    mutate(spot_id = gsub(paste0(id, "_"), "", spot_id)) %>%
    select(c("spot_id", paste0("q05cell_abundance_w_sf_", cellTypes))) %>%
    rename("Barcode" = "spot_id") %>%
    filter(Barcode %in% spotBarcodes)
  ### Order abundances to be in the same order as the spots in the colData
  order <- match(abundance_df$Barcode, spotBarcodes)
  abundance_df <- abundance_df[order,]
  ## Import abundances into colData -------------------------------------------#
  message("\tImporting to colData...")
  ### Merge with existing colData because otherwise it gets overwritten
  to_colData <- colData(msfe@sfe_data[[id]]) %>%
    as.data.frame() %>%
    left_join(abundance_df) %>%
    DataFrame()
  ### Import
  colData(msfe@sfe_data[[id]]) <- to_colData
  colnames(msfe@sfe_data[[id]]) <- spotBarcodes
}
## Housekeeping
rm(spotBarcodes, abundance_df, order, to_colData)

#### ----Build formulas --------------------------------------------------------
selectedCell_GWR <- c("AT1", "AT2", "Fibroblasts", "Macrophages", "MUC5B.",
                      "KRT5..KRT17.", "Transitional.AT2", "Endothelial.Cells",
                      "Myofibroblasts", "HAS1.High.Fibroblasts")
selectedCell_GWR <- paste0("q05cell_abundance_w_sf_", selectedCell_GWR)
selectedCells_GWR <- tidyr::expand_grid(selectedCell_GWR, selectedCell_GWR)
## Remove self pairs and duplicate pairs
selectedCells_GWR <- selectedCells_GWR %>%
  filter(selectedCell_GWR...1 != selectedCell_GWR...2) %>%
  # Create a temporary column that holds sorted pairs
  mutate(pair = pmin(selectedCell_GWR...1, selectedCell_GWR...2),
         pair2 = pmax(selectedCell_GWR...1, selectedCell_GWR...2)) %>%
  # Remove duplicate pairs
  distinct(pair, pair2, .keep_all = TRUE) %>%
  # Drop the temporary columns
  select(-pair, -pair2)
## Generate the formulas
formulas_cells <- paste(selectedCells_GWR$selectedCell_GWR...1,
                        selectedCells_GWR$selectedCell_GWR...2,
                        sep = "~")
names(formulas_cells) <- gsub("q05cell_abundance_w_sf_", "", formulas_cells)

#### ----Run GWR----------------------------------------------------------------
gwr_list_cell <- vector(mode = "list", length = length(sampleNames_selected))
gwr_runs_cell <- vector(mode = "list", length = length(formulas_cells))
names(gwr_list_cell) <- sampleNames_selected
names(gwr_runs_cell) <- names(formulas_cells)

for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  for (i in seq_along(formulas_cells)) {
    ## Some formulas return `Error: inv(): matrix is singular`. This is because
    ## there are too few values in the area to generate a regression model.
    ## For the sake of running everything once and tackling the erroring formulas
    ## one by one, we wrap everything inside a tryCatch function.
    tryCatch({
      message("\tWorking on formula: ", names(formulas_cells[i]), " (", i, "/", length(formulas_cells), ")")

      ### ----Set Bandwidth ---------------------------------------------------#
      message("\t\tSetting bandwidth...")
      spot_diameter <- msfe@sfe_data[[id]]@metadata[["spotDiameter"]][[id]][["spot_diameter_fullres"]]
      bw <- 3*spot_diameter


      ### ----Run GWR ---------------------------------------------------------#
      message("\t\tRunning GWR...")
      gwr_runs_cell[[names(formulas_cells[i])]] <- gwrSTE(gwr_method = "basic",
                                                          formula = formulas_cells[[i]],
                                                          m_sfe = msfe,
                                                          sample_id = id,
                                                          bw = bw,
                                                          kernel = "exponential")
    }, error = error)
  }

  ## Update output list
  message("Updating the list of GWR runs per sample.")
  gwr_list_cell[[id]] <- gwr_runs_cell
}

## Housekeeping
rm(spot_diameter, gwr_runs_cell)

#### ----Plot GWR --------------------------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Select GWR run for the specific sample
  gwrRuns <- gwr_list_cell[[id]]

  for (i in seq_along(formulas_cells)) {
    message("\tWorking on formula: ", names(formulas_cells[i]), " (", i, "/", length(formulas_cells), ")")

    ## Create folder
    graphics_out <- paste0(projectFolder, "/graphics_out/GWR_cellTypes/", names(formulas_cells[i]), "/")
    dir.create(graphics_out, recursive = TRUE)

    ## Plot GWR coefficients
    plotGWR(gwrRuns[[i]])

    prfx <- id
    main <- "_GWR"
    sfx <- paste0("_", names(formulas_cells[i]))
    other <- ""

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = grDevices::dev.size(units = "in")[1],
           height = grDevices::dev.size(units = "in")[2],
           units = "in",
           dpi = 300)
  }
  ## Housekeeping
  rm(gwRuns)
}


# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(gwr_list_cell, file = paste0(projectFolder, "/gwr_list_cell.rds"))

### ----Gene VS Cell -----------------------------------------------------------
#### ----Build formulas --------------------------------------------------------
selectedCell_GWR <- c("AT2", "Fibroblasts", "Macrophages", "MUC5B.",
                      "KRT5..KRT17.", "Transitional.AT2", "Endothelial.Cells",
                      "Myofibroblasts", "HAS1.High.Fibroblasts",
                      "MUC5AC..High", "SCGB3A2..SCGB1A1.", "SCGB3A2.",
                      "Proliferating.Epithelial.Cells")
selectedCell_GWR <- paste0("q05cell_abundance_w_sf_", selectedCell_GWR)
selectedGenes_GWR <- c(senescenceMarkers_ENSG, otherGenes_ENSG)
selectedCellGenes_GWR <- tidyr::expand_grid(selectedCell_GWR, selectedGenes_GWR)

## Generate the formulas
formulas_cellGenes <- paste(selectedCellGenes_GWR$selectedCell_GWR,
                            selectedCellGenes_GWR$selectedGenes_GWR,
                            sep = "~")
names(formulas_cellGenes) <- paste(selectedCellGenes_GWR$selectedCell_GWR,
                                   names(selectedGenes_GWR),
                                   sep = "~") %>%
  gsub("q05cell_abundance_w_sf_", "", .)

#### ----Run GWR----------------------------------------------------------------
gwr_list_cellGenes <- vector(mode = "list", length = length(sampleNames_selected))
gwr_runs_cellGenes <- vector(mode = "list", length = length(formulas_cellGenes))
names(gwr_list_cellGenes) <- sampleNames_selected
names(gwr_runs_cellGenes) <- names(formulas_cellGenes)

for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  for (i in seq_along(formulas_cellGenes)) {
    ## Some formulas return `Error: inv(): matrix is singular`. This is because
    ## there are too few values in the area to generate a regression model.
    ## For the sake of running everything once and tackling the erroring formulas
    ## one by one, we wrap everything inside a tryCatch function.
    tryCatch({
      message("\tWorking on formula: ", names(formulas_cellGenes[i]), " (", i, "/", length(formulas_cellGenes), ")")

      ### ----Set Bandwidth ---------------------------------------------------#
      message("\t\tSetting bandwidth...")
      spot_diameter <- msfe@sfe_data[[id]]@metadata[["spotDiameter"]][[id]][["spot_diameter_fullres"]]
      bw <- 3*spot_diameter


      ### ----Run GWR ---------------------------------------------------------#
      message("\t\tRunning GWR...")
      gwr_runs_cellGenes[[names(formulas_cellGenes[i])]] <- gwrSTE(gwr_method = "basic",
                                                                   formula = formulas_cellGenes[[i]],
                                                                   m_sfe = msfe,
                                                                   sample_id = id,
                                                                   bw = bw,
                                                                   kernel = "exponential")
    }, error = error)
  }

  ## Update output list
  message("Updating the list of GWR runs per sample.")
  gwr_list_cellGenes[[id]] <- gwr_runs_cellGenes
}

## Housekeeping
rm(spot_diameter, gwr_runs_cellGenes)

#### ----Plot GWR --------------------------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Select GWR run for the specific sample
  gwrRuns <- gwr_list_cellGenes[[id]]

  for (i in seq_along(formulas_cellGenes)) {
    message("\tWorking on formula: ", names(formulas_cellGenes[i]), " (", i, "/", length(formulas_cellGenes), ")")

    ## Create folder
    graphics_out <- paste0(projectFolder, "/graphics_out/GWR_geneVScell/", names(formulas_cellGenes[i]), "/")
    dir.create(graphics_out, recursive = TRUE)

    ## Plot GWR coefficients
    plotGWR(gwrRuns[[i]])

    prfx <- id
    main <- "_GWR"
    sfx <- paste0("_", names(formulas_cellGenes[i]))
    other <- ""

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = grDevices::dev.size(units = "in")[1],
           height = grDevices::dev.size(units = "in")[2],
           units = "in",
           dpi = 300)
  }
  ## Housekeeping
  rm(gwRuns)
}

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(gwr_list_cellGenes, file = paste0(projectFolder, "/gwr_list_cellGenes.rds"))

# ---------------------------------------------------------------------------- #
## ----Senescence Score Analysis -----------------------------------------------
### ----Set up new MSFE---------------------------------------------------------
msfe_gsva <- MetaSpatialFeatureExperiment()

### ----Generate senescence score ----------------------------------------------
#### ----Create gene sets to use ----------------------#
#### A) A selection of 4 senescence markers
sen_score_gs_selection <- selectedGenes_ENSG[c("CDKN1A", "GLB1", "CCL2", "IL6")]

#### B) A subset from the senMayo dataset with the genes that are most present
####    in the dataset.
sen_score_gs_senM <- c(senMayoGenes_ENSG,
                       selectedGenes_ENSG[c("CCL2", "IL6", "IL6ST", "GDF15")])

# Combine all genes across datasets
all_genes <- Reduce(intersect, lapply(msfe@sfe_data, rownames))

# Filter the 125 genes that are common across all datasets
genes_to_analyse <- intersect(all_genes, sen_score_gs_senM)

# Calculate presence proportions for each dataset
presence_list <- lapply(msfe@sfe_data,
                        calculate_presence_proportion,
                        genes = genes_to_analyse,
                        threshold = 0)

# Combine the presence proportions into a data frame
presence_df <- do.call(cbind, presence_list) %>% data.frame()
colnames(presence_df) <- c("Healthy", "IPF1", "IPF2", "IPF3")
rownames(presence_df) <- names(selectedGenes_ENSG)[match(rownames(presence_df), selectedGenes_ENSG)]

# Plot senMayo gene proportions in each sample
column_annot <- data.frame(FS = c(0, 1, 2, 3))
rownames(column_annot) <- c("Healthy", "IPF1", "IPF2", "IPF3")
ph <- pheatmap(presence_df,
         show_rownames = TRUE,
         show_colnames = TRUE,
         cluster_rows = TRUE,
         cluster_cols = FALSE,
         main = "Proportion of SenMayo senescence markers",
         scale = "row",
         annotation_col = column_annot,
         angle_col = 0)

graphics_out <- paste0(projectFolder, "/graphics_out/Senescence/")
dir.create(graphics_out, recursive = TRUE)
prfx <- ""
main <- "senMayo_proportions"
sfx <- "_complete"
other <- "_std"

ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
       plot = ph,
       device = "png",
       width = 7.77,
       height = 14,
       units = "in",
       dpi = 300)

# Calculate the average presence proportion across the two most fibrotic samples
presence_df$AvProp_IPF3_4 <- rowMeans(presence_df[,3:4])

# Set a threshold for selecting genes (e.g., expressed in at least 30% of locations)
threshold <- 0.3
selected_genes <- rownames(presence_df)[presence_df$AvProp_IPF3_4 >= threshold]

# Plot senMayo gene proportions in each sample for the selected senMayo subset
column_annot <- data.frame(FS = c(0, 1, 2, 3))
rownames(column_annot) <- c("Healthy", "IPF1", "IPF2", "IPF3")
ph <- pheatmap(presence_df[selected_genes,-ncol(presence_df)],
               show_rownames = TRUE,
               show_colnames = TRUE,
               cluster_rows = TRUE,
               cluster_cols = FALSE,
               main = "Proportion of SenMayo senescence markers",
               scale = "row",
               annotation_col = column_annot,
               angle_col = 0)

graphics_out <- paste0(projectFolder, "/graphics_out/Senescence/")
dir.create(graphics_out, recursive = TRUE)
prfx <- ""
main <- "senMayo_proportions"
sfx <- "_30plus"
other <- "_std"

ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
       plot = ph,
       device = "png",
       width = 7.77,
       height = 7.2,
       units = "in",
       dpi = 300)

# Display the selected genes
print(selected_genes)


#### ----The GSVA input -------------------------------#
genesets <- list("senScoreSelect" = sen_score_gs_selection,
                 "senScoreSenMayo" = selectedGenes_ENSG[selected_genes])

#### ----Run GSVA -------------------------------------#
## Select samples to run
samples_gsva <- sampleNames_selected

## GSVA: Gene Set Variation Analysis
for (i in seq_along(samples_gsva)) {
  id <- samples_gsva[i]
  message("Working on sample: ", id, " (", i, "/", length(samples_gsva), ")")

  ## Construct parameters
  gsva_par <- gsvaParam(exprData = msfe@sfe_data[[id]],
                        assay = "logcounts",
                        geneSets = genesets)

  ## Run GSVA
  gsva_out <- gsva(gsva_par)

  ## It is returned as a SpatialExperiment thus missing a few things that we add back
  class(gsva_out) <- "SpatialFeatureExperiment"
  colGeometry(gsva_out, "spotHex") <- colGeometry(msfe@sfe_data[[id]], "spotHex")
  rowData(gsva_out)[["gene_name"]] <- rownames(gsva_out)
  colGraph(gsva_out, "visium") <- colGraph(msfe@sfe_data[[id]])

  ## Output data into the msfe object
  msfe_gsva <- addSFE(msfe_gsva, gsva_out, sample_id = id)
}


### ----Plot Senescence Scores -------------------------------------------------
for (i in seq_along(samples_gsva)) {
  id <- samples_gsva[i]
  message("Working on sample: ", id, " (", i, "/", length(samples_gsva), ")")

  ## Fetch senescence scores names
  senScores <- rownames(msfe_gsva@sfe_data[[id]])
  names(senScores) <- c("Selection", "senMayo")

  for (j in seq_along(senScores)) {
    s <- senScores[j]
    s_name <- names(s)
    message(paste0("\t... on gene: ", s_name, " (", j, "/", length(senScores), ")"))

    ## Create folder
    graphics_out <- paste0(projectFolder, "/graphics_out/Expression/SenScores/")
    dir.create(graphics_out, recursive = TRUE)

    plotGeneExpression(msfe_gsva,
                       sample_id = id,
                       genes = s,
                       assay = "es",
                       title_type = "") +
      cols4all::scale_fill_continuous_c4a_div(palette = "-c4a.pu_gn_div")

    prfx <- id
    main <- "_SenScore"
    sfx <- paste0("_", s_name)
    other <- ""

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = 5.22,
           height = 8.23,
           units = "in",
           dpi = 300)
  }
  ## Housekeeping
  rm(senScores, s_name)
}

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(msfe_gsva, file = paste0(projectFolder, "/msfe_gsva.rds"))

### ---- Spatial Autocorrelation -----------------------------------------------
#### ----Local Moran's I -------------------------------------------------------
## Run Local statistic only for the genes that where selected from the global
## statistic
for (id in samples_gsva) {
  message("Working on sample: ", id)

  msfe_gsva <- moranLocalIPerm(m_sfe = msfe_gsva,
                               sample_id = id,
                               genes = rownames(msfe_gsva@sfe_data[[id]]),
                               assay = "es",
                               mc.cores = 6)
}


#### ----Local Geary's C -------------------------------------------------------
## Run Local statistic only for the genes that where selected from the global
## statistic
for (id in samples_gsva) {
  message("Working on sample: ", id)

  msfe_gsva <- gearyLocalCPerm(m_sfe = msfe_gsva,
                               sample_id = id,
                               genes = rownames(msfe_gsva@sfe_data[[id]]),
                               assay = "es",
                               mc.cores = 6)
}


#### ----Local Getis & Ord's G -------------------------------------------------
## Run Local statistic only for the genes that where selected from the global
## statistic
for (id in samples_gsva) {
  message("Working on sample: ", id)

  msfe_gsva <- getisLocalGPerm(m_sfe = msfe_gsva,
                               sample_id = id,
                               genes = rownames(msfe_gsva@sfe_data[[id]]),
                               assay = "es",
                               mc.cores = 6)
}

#### ----Plot local SA ----------------------------------------------------------
for (j in seq_along(samples_gsva)) {
  id <- samples_gsva[j]
  message("Working on sample: ", id, " (", j, "/", length(samples_gsva), ")")

  ## Fetch scores
  senScores <- rownames(msfe_gsva@sfe_data[[id]])

  for (s in senScores) {
    message(paste0("\t... on score: ", s))

    for (t in c("moran", "geary", "getis")) {
      if (t == "moran") {
        message("\t\t... on Moran's I")
        graphics_out <- paste0(projectFolder, "/graphics_out/SA/SA_Local_MoranI/", s, "/")
        dir.create(graphics_out, recursive = TRUE)
        main <- "_SA_LocalI"

      } else if (t == "geary") {
        message("\t\t... on Geary's C")
        graphics_out <- paste0(projectFolder, "/graphics_out/SA/SA_Local_GearyC/", s, "/")
        dir.create(graphics_out, recursive = TRUE)
        main <- "_SA_LocalC"

      } else if (t == "getis") {
        message("\t\t... on Getis & Ord's G")
        graphics_out <- paste0(projectFolder, "/graphics_out/SA/SA_Local_GetisG/", s, "/")
        dir.create(graphics_out, recursive = TRUE)
        main <- "_SA_LocalG"
      }

      ## Plot
      for (l in c("all", "significant")) {
        p1 <- plotSA_local(m_sfe = msfe_gsva,
                           sample_id = id,
                           feature = s,
                           statistic = t,
                           test = "permutation",
                           pVal = 0.05,
                           type = "hex",
                           title = "name",
                           locations = l)

        p2 <- plotSA_localClust(m_sfe = msfe_gsva,
                                sample_id = id,
                                feature = s,
                                statistic = t,
                                test = "permutation",
                                pVal = 0.05,
                                type = "hex",
                                title = "name",
                                clust_col = NULL,
                                locations = l)

        print((p1 / p2))

        prfx <- id
        main <- main
        sfx <- paste0("_", s)
        other <- ifelse(l == "all", "_all", "_sig")

        ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
               device = "png",
               width = 5.22,
               height = 8.23,
               units = "in",
               dpi = 300)
      }
    }
  }
}
## Housekeeping
rm(senScores, p1, p2)

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(msfe_gsva, file = paste0(projectFolder, "msfe_gsva.rds"))


### ----Subset - opposite SenMayo score (positive VS negative areas) -----------
## Compare different variables between areas with positive and negative SenMayo
## enrichment using violin/ rain-cloud plots.
sen_cutoff <- 0.3
cell_types <- cellTypes
genes <- selectedGenes_ENSG[sub_genes]
names(genes) <- sub_genes

for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Create folder
  graphics_out <- paste0(projectFolder, "/graphics_out/Combo/subset_senMayo/")
  dir.create(graphics_out, recursive = TRUE)

  ## Extract senMayo enrichment scores
  data_senMayo <- assay(msfe_gsva@sfe_data[[id]], "es")["senScoreSenMayo",] %>%
    data.frame() %>%
    rename("senMayo_score" = ".") %>%
    mutate(sen_group = case_when(senMayo_score > sen_cutoff ~ "positive",
                                 senMayo_score < -sen_cutoff ~ "negative",
                                 .default = "no"),
           Barcode = rownames(.))
  data_senMayo_cols_to_use <- c("sen_group", "Barcode")

  ## Extract cell type abundances
  data_cellAb <- colData(msfe@sfe_data[[id]]) %>%
    data.frame() %>%
    select(starts_with("q05cell")) %>%
    select(contains(cell_types)) %>%
    mutate(Barcode = rownames(.))

  ## Fix column names
  colnames(data_cellAb) <- gsub("q05cell_abundance_w_sf_", "", colnames(data_cellAb))

  ## Merge and pivot longer for faceting
  data <- left_join(data_senMayo[data_senMayo_cols_to_use],
                    data_cellAb,
                    by = "Barcode") %>%
    select(-"Barcode") %>%
    pivot_longer(cols = -c("sen_group"),
                 names_to = "cell_type",
                 values_to = "cell_density") %>%
    filter(sen_group != "no")

  data$cell_type <- factor(data$cell_type,
                           levels = cellTypes,
                           labels = names(cellTypes))

  ## Plot SenMayo score for locations passing the cutoff only -----------------#
  geoms <- colGeometry(msfe_gsva@sfe_data[[id]], "spotHex") %>%
    data.frame() %>%
    mutate(Barcode = colnames(msfe_gsva@sfe_data[[id]]))
  rownames(geoms) <- colnames(msfe_gsva@sfe_data[[id]])

  data_senMayo <- left_join(data_senMayo, geoms, by = "Barcode") %>%
    filter(sen_group != "no")

  image <- image_to_plot_lst[[id]][["image"]]
  max_col_value <- image_to_plot_lst[[id]][["max_col_value"]]
  limits_list <- image_to_plot_lst[[id]][["limits_list"]]

  ggplot(data_senMayo,
         aes(geometry = geometry)) +
    tidyterra::geom_spatraster_rgb(data = image,
                                   max_col_value = max_col_value,
                                   alpha = 0.1) +
    # ggplot2::lims(x = limits_list[[2]],
    #               y = limits_list[[1]]) +
    geom_sf(aes(fill = senMayo_score)) +
    scale_fill_continuous_c4a_div(reverse = TRUE) +
    theme_void()

  prfx <- id
  main <- ""
  sfx <- "_senMayoScores_"
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 7.45,
         height = 5.45,
         units = "in",
         dpi = 300)

  ## Plot cell Abundances in rain-cloud plots ---------------------------------#
  message("\tPlotting cell abundance rain-clouds...")
  ggplot(data,
         aes(x = sen_group, y = cell_density, fill = sen_group)) +
    ggdist::stat_halfeye(adjust = 0.5, justification = -0.2, .width = 0, point_colour = NA) +
    geom_boxplot(width = 0.1, alpha = 0.5, outlier.color = NA) +
    ggdist::stat_dots(side = "left", justification = 1.1, binwidth = NA, overflow = "compress") +
    scale_fill_manual(values = c("#246C27", "#762A83")) +
    # stat_summary(fun.data = mean_cl_boot, geom = "errorbar", width = 0.2, colour = "black") +  # Add mean with bootstrapped confidence interval
    # stat_summary(fun = mean, geom = "point", shape = 23, size = 2, fill = "white") + # Add mean point
    ggpubr::stat_compare_means(ref.group = "negative",
                               method = "wilcox.test",
                               label = "p.signif",
                               tip.length = 0.005) +
    labs(title = "",
         y = "Absolute Cell Density",
         x = "",
         fill = "SenMayo\nenrichment\nscore") +
    facet_wrap(~cell_type, nrow = 3) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
          axis.text.y = element_text(size = 15),
          axis.title.y = element_text(size = 20),
          legend.text = element_text(size = 15),
          legend.title = element_text(size = 20))

  message("\tSaving ...")
  prfx <- id
  main <- ""
  sfx <- "_rainCloud_"
  other <- "perGroup_CellDensities"

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 15,
         height = 12,
         units = "in",
         dpi = 300)

  ## Plot gene expression -----------------------------------------------------#
  ## Merge and pivot longer for expression
  data_expr <- assay(msfe@sfe_data[[id]], "logcounts")[genes,] %>%
    t() %>%
    data.frame() %>%
    mutate(Barcode = rownames(.))

  colnames(data_expr) <- c(names(genes), "Barcode")

  data <- data_senMayo %>%
    select(-c("senMayo_score", "geometry")) %>%
    left_join(., data_expr, by = "Barcode") %>%
    select(-"Barcode") %>%
    pivot_longer(cols = -c("sen_group"),
                 names_to = "gene",
                 values_to = "logcounts") %>%
    filter(sen_group != "no")

  ## Plot
  message("\tPlotting gene expression rain-clouds...")
  ggplot(data,
         aes(x = sen_group, y = logcounts, fill = sen_group)) +
    ggdist::stat_halfeye(adjust = 0.5, justification = -0.2, .width = 0, point_colour = NA) +
    geom_boxplot(width = 0.1, alpha = 0.5, outlier.color = NA) +
    ggdist::stat_dots(side = "left", justification = 1.1, binwidth = NA, overflow = "compress") +
    scale_fill_manual(values = c("#246C27", "#762A83")) +
    # stat_summary(fun.data = mean_cl_boot, geom = "errorbar", width = 0.2, colour = "black") +  # Add mean with bootstrapped confidence interval
    # stat_summary(fun = mean, geom = "point", shape = 23, size = 2, fill = "white") + # Add mean point
    ggpubr::stat_compare_means(ref.group = "negative",
                               method = "wilcox.test",
                               label = "p.signif",
                               tip.length = 0.005) +
    labs(title = "",
         y = "Normalised Log2-counts",
         x = "",
         fill = "SenMayo\nenrichment\nscore") +
    facet_wrap(~gene, nrow = 2) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
          axis.text.y = element_text(size = 20),
          axis.title.y = element_text(size = 20),
          legend.text = element_text(size = 15),
          legend.title = element_text(size = 20))

  message("\tSaving ...")
  prfx <- id
  main <- ""
  sfx <- "_rainCloud_"
  other <- "perGroup_GeneExpression"

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 10.83,
         height = 8.83,
         units = "in",
         dpi = 300)
}


### ---- GWR -------------------------------------------------------------------
#### ----Build formulas --------------------------------------------------------
selectedCell_GWR <- paste0("q05cell_abundance_w_sf_",
                           c("AT2", "Fibroblasts", "Macrophages", "MUC5B.", "AT1",
                             "KRT5..KRT17.", "Transitional.AT2", "Endothelial.Cells",
                             "Myofibroblasts", "HAS1.High.Fibroblasts",
                             "MUC5AC..High", "SCGB3A2..SCGB1A1.", "SCGB3A2.",
                             "Proliferating.Epithelial.Cells"))
senScores_GWR <- c("senScoreSelect", "senScoreSenMayo")

senScore_cell_pairs_GWR <- tidyr::expand_grid(senScores_GWR, selectedCell_GWR)

## Generate the formulas
formulas_senVSCells <- paste(senScore_cell_pairs_GWR$senScores_GWR,
                            senScore_cell_pairs_GWR$selectedCell_GWR,
                            sep = "~")
names(formulas_senVSCells) <- gsub("q05cell_abundance_w_sf_", "", formulas_senVSCells)

## Add some more formulas to test
cell_pairs_GWR <- c("q05cell_abundance_w_sf_Endothelial.Cells+q05cell_abundance_w_sf_Myofibroblasts",
                    "q05cell_abundance_w_sf_Proliferating.Epithelial.Cells+q05cell_abundance_w_sf_Myofibroblasts",
                    "q05cell_abundance_w_sf_AT2+q05cell_abundance_w_sf_Myofibroblasts",
                    "q05cell_abundance_w_sf_AT2+q05cell_abundance_w_sf_AT1")
senScore_cell_pairs_GWR2 <- tidyr::expand_grid(senScores_GWR, cell_pairs_GWR)
formulas_senVSCells2 <- paste(senScore_cell_pairs_GWR2$senScores_GWR,
                             senScore_cell_pairs_GWR2$cell_pairs_GWR,
                             sep = "~")
names(formulas_senVSCells2) <- gsub("q05cell_abundance_w_sf_", "", formulas_senVSCells2)

## Update formulas vector
formulas_senVSCells <- c(formulas_senVSCells, formulas_senVSCells2)

#### ----Run GWR ---------------------------------------------------------------
gwr_list_senScVsCell <- vector(mode = "list", length = length(sampleNames_selected))
gwr_runs_senScVsCell <- vector(mode = "list", length = length(formulas_senVSCells))
names(gwr_list_senScVsCell) <- sampleNames_selected
names(gwr_runs_senScVsCell) <- names(formulas_senVSCells)

for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  for (i in seq_along(formulas_senVSCells)) {
    ## Some formulas return `Error: inv(): matrix is singular`. This is because
    ## there are too few values in the area to generate a regression model.
    ## For the sake of running everything once and tackling the erroring formulas
    ## one by one, we wrap everything inside a tryCatch function.
    f <- formulas_senVSCells[i]
    tryCatch({
      message("\tWorking on formula: ", names(f), " (", i, "/", length(formulas_senVSCells), ")")

      ### ----Set Bandwidth ---------------------------------------------------#
      message("\t\tSetting bandwidth...")
      spot_diameter <- msfe@sfe_data[[id]]@metadata[["spotDiameter"]][[id]][["spot_diameter_fullres"]]
      bw <- 3*spot_diameter


      ### ----Run GWR ---------------------------------------------------------#
      message("\t\tRunning GWR...")

      gwr_runs_senScVsCell[[names(f)]] <- gwrSTE(gwr_method = "basic",
                                                 formula = f,
                                                 m_sfe = msfe_gsva,
                                                 sample_id = id,
                                                 bw = bw,
                                                 kernel = "exponential",
                                                 assay = "es")
    }, error = error)
  }

  ## Update output list
  message("Updating the list of GWR runs per sample.")
  gwr_list_senScVsCell[[id]] <- gwr_runs_senScVsCell
}

## Housekeeping
rm(spot_diameter, gwr_runs_senScVsCell)



#### ----Plot GWR --------------------------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Select GWR run for the specific sample
  gwrRuns <- gwr_list_senScVsCell[[id]]

  for (i in seq_along(formulas_senVSCells)) {
    f <- formulas_senVSCells[i]
    message("\tWorking on formula: ", names(f), " (", i, "/", length(formulas_senVSCells), ")")

    ## Create folder
    graphics_out <- paste0(projectFolder, "/graphics_out/GWR_senScVScell/", names(f), "/")
    dir.create(graphics_out, recursive = TRUE)

    ## Get independent variables
    independent_vars <- strsplit(f, "~")[[1]][2] %>%
      strsplit(., "\\+") %>% .[[1]]

    ## Plot GWR coefficients
    plotGWR(gwrRuns[[i]], predictorVar = independent_vars, title = names(f))

    prfx <- id
    main <- "_GWR"
    sfx <- paste0("_", names(f))
    other <- ""

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = grDevices::dev.size(units = "in")[1],
           height = grDevices::dev.size(units = "in")[2],
           units = "in",
           dpi = 300)

    ## Plot local R^2
    plotGWR_R2(gwrRuns[[i]])

    prfx <- id
    main <- "_GWR"
    sfx <- paste0("_", names(f))
    other <- "_R2"

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = grDevices::dev.size(units = "in")[1],
           height = grDevices::dev.size(units = "in")[2],
           units = "in",
           dpi = 300)

  }
  ## Housekeeping
  rm(gwRuns)
}

# ---------------------------------------------------------------------------- #
## ----Checkpoint ------------------------------------------------------------ #
saveRDS(gwr_list_senScVsCell, file = paste0(projectFolder, "/gwr_list_senScVsCell.rds"))

### ----Subset - opposite AT2/Myofibroblast regression coef --------------------
## Here we will plot gene expression, cell abundance and SA results only for
## areas where AT2 and Myofibroblasts are
#### ----Select Variables ------------------------------------------------------
sub_genes <- c("IL6", "CDKN1A", "ACTA2", "COL1A1", "GLB1", "CCL2", "ICAM1",
               "SNAI1", "SNAI2", "TWIST1", "ZEB1", "ZEB2")
sub_genesENSG <- selectedGenes_ENSG[sub_genes]
sub_formulas <- c("senScoreSenMayo~Myofibroblasts",
                  "senScoreSenMayo~AT2",
                  "senScoreSenMayo~AT2+Myofibroblasts",
                  "senScoreSenMayo~Endothelial.Cells")
# sub_formulas <- formulas_senVSCells[sub_formulas]
sub_samples <- sampleNames_selected
sub_cell_names <- c("AT2", "Myofibroblasts", "Endothelial cells")
sub_cells <- paste0("q05cell_abundance_w_sf_", selectedCellTypes[sub_cell_names])
names(sub_cells) <- sub_cell_names

data_to_plot_lst <- vector(mode = "list", length = length(sub_samples))
names(data_to_plot_lst) <- sub_samples

image_to_plot_lst <- vector(mode = "list", length = length(sub_samples))
names(image_to_plot_lst) <- sub_samples

#### ----Prepare data ----------------------------------------------------------
for (j in seq_along(sub_samples)) {
  id <- sub_samples[j]
  message("Working on sample: ", id, " (", j, "/", length(sub_samples), ")")

  ## Select GWR run for the specific sample
  message("\tExtracting GWR data...")
  gwr_list <- gwr_list_senScVsCell[[id]]

  gwr_dt_list <- lapply(sub_formulas,
                        function(x){
                          ### Get predictor variable
                          mdl <- gwr_list[[x]]$GW.arguments$formula
                          predictorVar <- gsub(".+?~", "", mdl) %>%
                            strsplit(., "[+]")
                          predictorVar <- predictorVar[[1]]
                          ### Identify statistically significant areas
                          predTVcol <- paste0(predictorVar, "_TV")
                          dt <- gwr_toSF(gwr_list[[x]]) %>%
                            select(all_of(c(predictorVar, predTVcol,
                                            "Local_R2", "geometry"))) %>%
                            rownames_to_column(var = "Barcode") %>%
                            data.frame()
                          predictorVar <- ifelse(length(predictorVar) > 1,
                                                 paste(predictorVar, collapse = '+'),
                                                 predictorVar)
                          rename_with(dt, ~ paste0(predictorVar, "_R2", recycle0 = TRUE),
                                      all_of("Local_R2"))
                        })

  message("\tFiltering GWR results...")
  ## Filter for areas where senMayo~AT2 is negative and senMayo~Myofibroblasts is positive
  gwr_dt <- gwr_dt_list %>%
    reduce(left_join, by = c("Barcode", "geometry"), suffix = c("", "_ATMyo"))
  colnames(gwr_dt) <- gsub("q05cell_abundance_w_sf_", "", colnames(gwr_dt))
  gwr_dt <- gwr_dt %>%
    filter(AT2 < -0.1 & Myofibroblasts > 0.1)

  message("\tFetching gene expression data...")
  ### Fetch gene expression
  geneExpression <- SummarizedExperiment::assay(msfe@sfe_data[[id]], "logcounts")[sub_genesENSG, gwr_dt$Barcode] %>%
    t() %>%
    data.frame() %>%
    rownames_to_column(var = "Barcode")
  colnames(geneExpression) <- c("Barcode", names(sub_genesENSG))

  message("\tFetching SenMayo score data...")
  ### Fetch senescence scores
  senMayoScores <- SummarizedExperiment::assay(msfe_gsva@sfe_data[[id]], "es")["senScoreSenMayo", gwr_dt$Barcode] %>%
    #t() %>%
    data.frame() %>%
    rename("senMayo" = ".") %>%
    rownames_to_column(var = "Barcode")

  message("\tFetching cell abundance data...")
  ### Fetch cell abundance
  cell_abund <- cellAbundance[[id]][c("spot_id", sub_cells)] %>%
    rename("Barcode" = "spot_id")
  cell_abund$Barcode <- gsub(paste0(id, "_"), "", cell_abund$Barcode)
  cell_abund <- cell_abund[cell_abund$Barcode %in% gwr_dt$Barcode, ]

  message("\tFetching annotation data...")
  ### Fetch annotations
  spot_annot <- colData(msfe@sfe_data[[id]])["annotation"] %>%
    data.frame() %>%
    rownames_to_column(var = "Barcode")
  spot_annot <- spot_annot[spot_annot$Barcode %in% gwr_dt$Barcode, ]

  message("\tFetching image data...")
  ### Fetch Image data
  image <- imgRaster(getImg(msfe@sfe_data[[id]],
                            sample_id = id,
                            image_id = "lowres"))
  max <- image@ptr$range_max[1]
  if (is.na(max)) {
    max <- max(image[,,1])
  }
  if (max <= 1) {
    max_col_value <- 1
  } else if (max <= 255) {
    max_col_value <- 255
  } else if (max <= 65536) {
    max_col_value <- 65536
  }
  coords <- spatialCoords(msfe@sfe_data[[id]])[gwr_dt$Barcode,] %>%
    matrix(ncol = 2)
  limits_list <- list(y_lims = c(min(coords[,2]), max(coords[,2])),
                      x_lims = c(min(coords[,1]), max(coords[,1])))

  message("\tMerging data...")
  ### Merge expression and cell abundance with GWR output
  data_to_plot <- left_join(gwr_dt, geneExpression, by = "Barcode") %>%
    left_join(cell_abund, by = "Barcode", suffix = c("", ".y")) %>%
    left_join(spot_annot, by = "Barcode") %>%
    left_join(senMayoScores, by = "Barcode")
  colnames(data_to_plot) <- gsub("q05cell_abundance_w_sf_", "abundance_", colnames(data_to_plot))

  message("\tOutputing to list...")
  ## Output to the lists
  data_to_plot_lst[[id]] <- data_to_plot
  image_to_plot_lst[[id]] <- list(image = image,
                                  max_col_value = max_col_value,
                                  limits_list = limits_list)
}

#### ----Plot coefficients -----------------------------------------------------
## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/Combo/senMayo_AT2_Myof/")
dir.create(graphics_out, recursive = TRUE)


## Named colour values for consistent colouring of annotation labels
colours <- c("Blood Vessel" = "#1F78C8",
             "Large Airway" = "#2C3E50",
             "Normal Alveolar and Other" = "#33A02C",
             "Inflammation" = "#FFD700",
             "Suspect Fibrosis" = "#FD8D3C",
             "Diseased" = "#CC4C00")


for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Export data from lists
  message("\tExporting data from list...")
  data_to_plot <- data_to_plot_lst[[id]]
  image <- image_to_plot_lst[[id]][["image"]]
  max_col_value <- image_to_plot_lst[[id]][["max_col_value"]]
  limits_list <- image_to_plot_lst[[id]][["limits_list"]]

  message("\tPlotting...")
  ### Master plot
  p <- ggplot(data_to_plot,
              aes(geometry = geometry)) +
    tidyterra::geom_spatraster_rgb(data = image,
                                   max_col_value = max_col_value,
                                   alpha = 0.1) +
    ggplot2::lims(x = limits_list[[2]],
                  y = limits_list[[1]]) +
    theme_void()

  ### Plot GWR coefficient ----------#
  plot_gwr <- vector(mode = "list", length = length(sub_cell_names))
  names(plot_gwr) <- sub_cell_names

  for (c in sub_cell_names) {
    lg <- paste0(c," ")
    plot_gwr[[c]] <- p +
      geom_sf(aes(fill = !!sym(c))) + # force evaluate aes
      scale_fill_gradient2(high = "#880A1F",
                           mid = "#FAFBF7",
                           low = "#426E92",
                           midpoint = 0,
                           n.breaks = 7) +
      labs(fill = bquote(.(lg) ~ beta[1]))
  }

  p1 <- patchwork::wrap_plots(plot_gwr, ncol = 1)

  ### Plot GWR T-values ----------#
  plot_tv <- vector(mode = "list", length = length(sub_cell_names))
  names(plot_tv) <- sub_cell_names

  for (c in sub_cell_names) {
    tv <- paste0(c,"_TV")
    plot_tv[[c]] <- p +
      geom_sf(aes(fill = !!sym(tv))) + # force evaluate aes
      scale_fill_gradient2(high = "#8C510A",
                           mid = "#F5F5F5",
                           low = "#01665E",
                           midpoint = 0,
                           n.breaks = 7) +
      labs(fill = tv)
  }

  p2 <- patchwork::wrap_plots(plot_tv, ncol = 1)

  ### Plot GWR local R2 ----------#
  plot_r2 <- vector(mode = "list", length = length(sub_cell_names))
  names(plot_r2) <- sub_cell_names

  for (c in sub_cell_names) {
    lg <- paste0(c, " ")
    r2 <- paste0(c,"_R2")
    plot_r2[[c]] <- p +
      geom_sf(aes(fill = !!sym(r2))) + # force evaluate aes
      scale_fill_viridis_c() +
      labs(fill = bquote(.(lg) ~ R^2))
  }

  p3 <- patchwork::wrap_plots(plot_r2, ncol = 1)

  ### Plot cell abundance --------#
  plot_cellAbund <- vector(mode = "list", length = length(sub_cells))
  names(plot_cellAbund) <- names(sub_cells)

  for (c in sub_cell_names) {
    ab <- paste0("abundance_", c)
    plot_cellAbund[[c]] <- p +
      geom_sf(aes(fill = !!sym(ab))) + # force evaluate aes
      scale_fill_viridis_c(option = "turbo") +
      labs(fill = paste(c, "Density", sep = "\n"))
  }

  p4 <- patchwork::wrap_plots(plot_cellAbund, ncol = 1)

  ### Plot annotation ------------#
  # p5 <- p +
  #   geom_sf(aes(fill = annotation)) +
  #   scale_fill_manual(values = colours) + # colours as defined in the annotation plots earlier
  #   labs(fill = "Annotation")

  p1 | p2 | p3 | p4

  message("\tSaving ...")
  prfx <- id
  main <- "_GWR"
  sfx <- paste0("_", f)
  other <- "_subset_senMayo"

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 19.16,
         height = 8.25,
         units = "in",
         dpi = 300)

  ### Plot annotation ------------#
  p +
    geom_sf(aes(fill = annotation)) +
    scale_fill_manual(values = colours) + # colours as defined in the annotation plots earlier
    labs(fill = "Annotation")

  message("\tSaving ...")
  prfx <- id
  main <- "_GWR"
  sfx <- paste0("_", f)
  other <- "_subset_senMayo_annotation"

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 6.09,
         height = 5.41,
         units = "in",
         dpi = 300)
}

#### ----Plot gene expression --------------------------------------------------
## Set folder
graphics_out <- paste0(projectFolder, "/graphics_out/Combo/senMayo_AT2_Myof/")

for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Export data from lists
  message("\tExporting data from list...")
  data_to_plot <- data_to_plot_lst[[id]]
  image <- image_to_plot_lst[[id]][["image"]]
  max_col_value <- image_to_plot_lst[[id]][["max_col_value"]]
  limits_list <- image_to_plot_lst[[id]][["limits_list"]]

  message("\tPlotting...")
  ### Master plot
  p <- ggplot(data_to_plot,
              aes(geometry = geometry)) +
    tidyterra::geom_spatraster_rgb(data = image,
                                   max_col_value = max_col_value,
                                   alpha = 0.1) +
    # ggplot2::lims(x = limits_list[[2]],
    #               y = limits_list[[1]]) +
    theme_void()

  ### Plot Gene expression
  plot_geneExpr <- vector(mode = "list", length = length(sub_genesENSG))
  names(plot_geneExpr) <- names(sub_genesENSG)

  for (g in names(sub_genesENSG)) {
    plot_geneExpr[[g]] <- p +
      geom_sf(aes(fill = !!sym(g))) + # force evaluate aes
      scale_fill_viridis_c() +
      labs(fill = paste(g, "Norm. Log2\ncounts", sep = "\n"))
  }

  p1 <- patchwork::wrap_plots(plot_geneExpr, nrow = 3, ncol = 3)

  ### Plot SenMayo senescence enrichment score
  p2 <- p +
    geom_sf(aes(fill = senMayo)) +
    scale_fill_gradient2(high = "#762A83",
                         mid = "#E8E8E8",
                         low = "#246C27",
                         midpoint = 0,
                         n.breaks = 7) +
    labs(fill = "SenMayo\nenrichment\nscore")

  # c("#83488B", "#C082C7", "#E5C5EA", "#E8E8E8", "#B4D8B5", "#61A863", "#246C27")
  # c("#762A83", "#AF8DC3", "#E7D4E8", "#F7F7F7", "#D9F0D3", "#7FBF7B", "#1B7837")
  # c("#C51B7D", "#E9A3C9", "#FDE0EF", "#F7F7F7", "#E6F5D0", "#A1D76A", "#4D9221")

  p1 | p2

  message("\tSaving ...")
  prfx <- id
  main <- ""
  sfx <- ""
  other <- "_geneExpr_SenMayo"

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 19.16,
         height = 8.25,
         units = "in",
         dpi = 300)
}

#### ----Plot rain-cloud cell abundance and gene expression --------------------
## Set folder
graphics_out <- paste0(projectFolder, "/graphics_out/Combo/senMayo_AT2_Myof/")

for (j in seq_along(sampleNames_selected)[2:4]) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Plot cell abundance as rain-cloud plots ----------------------------------#
  ## Export cell type data from lists
  message("\tExporting data from list...")
  data_to_plot <- data_to_plot_lst[[id]] %>%
    select(matches("abund|senM")) %>%
    mutate(sen_group = case_when(senMayo > 0 ~ "positive",
                                 senMayo < 0 ~ "negative",
                                 .default = "no")) %>%
    pivot_longer(cols = -c("sen_group", "senMayo"),
                 names_to = "cell_type",
                 values_to = "cell_density") %>%
    filter(sen_group != "no") %>%
    mutate(cell_type = gsub("abundance_|.Cells", "", .data$cell_type))

  message("\tPlotting rain-cloud Cell density...")
  ggplot(data_to_plot,
         aes(x = sen_group, y = cell_density, fill = sen_group)) +
    ggdist::stat_halfeye(adjust = 0.5, justification = -0.2, .width = 0, point_colour = NA) +
    geom_boxplot(width = 0.1, alpha = 0.5, outlier.color = NA) +
    # ggdist::stat_dots(side = "left", justification = 1.1, binwidth = NA, overflow = "compress") +
    scale_fill_manual(values = c("#246C27", "#762A83")) +
    ggpubr::stat_compare_means(ref.group = "negative",
                               method = "wilcox.test",
                               label = "p.signif",
                               tip.length = 0.005) +
    labs(title = "Cell Densities per Group",
         y = "Absolute Cell Density",
         x = "",
         fill = "SenMayo\nenrichment\nscore") +
    facet_wrap(~cell_type, nrow = 1) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
          axis.text.y = element_text(size = 20),
          axis.title.y = element_text(size = 20),
          legend.text = element_text(size = 15),
          legend.title = element_text(size = 20))

  message("\tSaving ...")
  prfx <- id
  main <- ""
  sfx <- "_rainCloud_"
  other <- "perGroup_CellDensities"

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 10.83,
         height = 6.83,
         units = "in",
         dpi = 300)
}

for (j in seq_along(sampleNames_selected)[2:4]) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Plot gene expression as rain-cloud plots ---------------------------------#
  ## Export cell type data from lists
  message("\tExporting data from list...")
  data_to_plot <- data_to_plot_lst[[id]] %>%
    select(all_of(c(sub_genes, "senMayo"))) %>%
    mutate(sen_group = case_when(senMayo > 0 ~ "positive",
                                 senMayo < 0 ~ "negative",
                                 .default = "no")) %>%
    pivot_longer(cols = -c("sen_group", "senMayo"),
                 names_to = "gene",
                 values_to = "logcounts") %>%
    filter(sen_group != "no")

  message("\tPlotting rain-cloud Gene Expression...")
  ggplot(data_to_plot,
         aes(x = sen_group, y = logcounts, fill = sen_group)) +
    ggdist::stat_halfeye(adjust = 0.5, justification = -0.2, .width = 0, point_colour = NA) +
    geom_boxplot(width = 0.1, alpha = 0.5, outlier.color = NA) +
    # ggdist::stat_dots(side = "left", justification = 1.1, binwidth = NA, overflow = "compress") +
    scale_fill_manual(values = c("#246C27", "#762A83")) +
    ggpubr::stat_compare_means(ref.group = "negative",
                               method = "wilcox.test",
                               label = "p.signif",
                               tip.length = 0.005) +
    labs(title = "",
         y = "Normalised Log2-counts",
         x = "",
         fill = "SenMayo\nenrichment\nscore") +
    facet_wrap(~gene, nrow = 1) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 20),
          axis.text.y = element_text(size = 20),
          axis.title.y = element_text(size = 20),
          legend.text = element_text(size = 15),
          legend.title = element_text(size = 20))

  message("\tSaving ...")
  prfx <- id
  main <- ""
  sfx <- "_rainCloud_"
  other <- "perGroup_GeneExpression"

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 10.83,
         height = 4.83,
         units = "in",
         dpi = 300)
}



# ---------------------------------------------------------------------------- #
## ----Additional Plots --------------------------------------------------------
### ----Annotation -------------------------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Create folder
  graphics_out <- paste0(projectFolder, "/graphics_out/Annotation/")
  dir.create(graphics_out, recursive = TRUE)

  ## Title
  title <- paste0(metadata$condition[metadata$sample_id == id],
                  " (",
                  metadata$subject_alias[metadata$sample_id == id],
                  ")")

  ## Plot Image and annotation
  p1 <- plotQC_tissueImg(msfe,
                         sample_id = id,
                         res = "lowres",
                         type = "none",
                         annotate = FALSE) +
    labs(title = title,
         subtitle = "") +
    theme(plot.title = element_text(hjust = 0.5))

  p2 <- plotQC_spotsAnnotation(msfe,
                               sample_id = id,
                               type = "hex",
                               colours = colours) +
    labs(title = "") # +
  # scale_fill_manual(limits = names(colours))

  print((p1 + p2))

  prfx <- id
  main <- "_Annotation"
  sfx <- ""
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 5.22,
         height = 8.23,
         units = "in",
         dpi = 300)

  ## Housekeeping
  rm(p1, p2, title)
}


### ----Gene expression of genes -----------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Plot Expression
  for (i in seq_along(selectedGenes_ENSG)) {
    g <- selectedGenes_ENSG[i]
    g_name <- names(g)
    message(paste0("\t... on gene: ", g_name, " (", i, "/", length(selectedGenes_ENSG), ")"))

    ## Create folder
    graphics_out <- paste0(projectFolder, "/graphics_out/Expression/", g_name, "/")
    dir.create(graphics_out, recursive = TRUE)

    plotGeneExpression(msfe,
                       sample_id = id,
                       genes = g,
                       assay = "logcounts")

    prfx <- id
    main <- "_Expression"
    sfx <- paste0("_", g_name)
    other <- ""

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = 5.22,
           height = 8.23,
           units = "in",
           dpi = 300)
  }
  ## Housekeeping
  rm(g_name)
}


### ----Cell Abundances --------------------------------------------------------
abundance_df_lst <- vector(mode = "list", length = length(sampleNames_selected))
names(abundance_df_lst) <- sampleNames_selected

#### ----Plot: map -------------------------------------------------------------
for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Extract abundances for specific cell types -------------------------------#
  abundance_df <- cellAbundance[[id]] %>%
    mutate(spot_id = gsub(paste0(id, "_"), "", spot_id)) %>%
    select(c("spot_id", paste0("q05cell_abundance_w_sf_", selectedCellTypes))) %>%
    rename("Barcode" = "spot_id")
  ## Filter out any abundance value less than 0.2 as it might be noise
  # abundance_df <- mutate(abundance_df, across(-Barcode, ~ifelse(. < 0.2, NA, .)))
  ## Fix column names
  colnames(abundance_df) <- gsub("q05cell_abundance_w_sf_", "", colnames(abundance_df))
  ## Extract geometries
  geoms <- colGeometry(msfe@sfe_data[[id]], "spotHex") %>%
    rownames_to_column(var = "Barcode")
  ## Merge abundances and geometries to plot
  abundance_df <- left_join(abundance_df, geoms, by = "Barcode")

  ## Output the list
  abundance_df_lst[[id]] <- abundance_df

  ## Plot a map per cell type -------------------------------------------------#
  for (cell in selectedCellTypes) {
    message(paste0("\t... on cell type: ", cell))

    ## Create folder
    graphics_out <- paste0(projectFolder, "/graphics_out/CellAbundance/", cell, "/")
    dir.create(graphics_out, recursive = TRUE)

    ## Prepare
    title <- gsub(".C", " c", cell)

    ## Plot
    ggplot(data = abundance_df,
           aes(fill = .data[[cell]])) +
      geom_sf(aes(geometry = geometry)) +
      scale_fill_viridis_c(option = "turbo",
                           na.value = "transparent") +
      labs(title = title,
           fill = "Cell\nabundance") +
      theme_void() +
      theme(plot.title = element_text(hjust = 0.5, size = 20))


    prfx <- id
    main <- "_CellAbundance"
    sfx <- paste0("_", cell)
    other <- ""

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           device = "png",
           width = 6.64,
           height = 4.59,
           units = "in",
           dpi = 300)
  }

  ## Plot as panel ------------------------------------------------------------#
  ## Create folder
  graphics_out <- paste0(projectFolder, "/graphics_out/CellAbundance/Panel/")
  dir.create(graphics_out, recursive = TRUE)

  abundance_df <- pivot_longer(abundance_df,
                               cols = -c("Barcode", "geometry"),
                               names_to = "cellType",
                               values_to = "abundance")

  ## Plot
  ggplot(data = abundance_df,
         aes(fill = abundance)) +
    geom_sf(aes(geometry = geometry)) +
    scale_fill_viridis_c() +
    labs(fill = "Cell\nabundance") +
    facet_wrap(~cellType) +
    theme_void()

  prfx <- id
  main <- "_CellAbundance"
  sfx <- "_panel"
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = 8.86,
         height = 7.43,
         units = "in",
         dpi = 300)

  ## Housekeeping
  rm(abundance_df, title, geoms)
}

#### ----Plot: heatmap ---------------------------------------------------------
graphics_out <- paste0(projectFolder, "/graphics_out/CellAbundance/Heatmaps/")
dir.create(graphics_out, recursive = TRUE)

for (j in seq_along(sampleNames_selected)) {
  id <- sampleNames_selected[j]
  message("Working on sample: ", id, " (", j, "/", length(sampleNames_selected), ")")

  ## Extract abundances for specific cell types -------------------------------#
  abundance_df <- abundance_df_lst[[id]] %>%
    column_to_rownames(var = "Barcode") %>%
    select(!all_of("geometry")) %>%
    t() %>%
    data.frame()
  colnames(abundance_df) <- gsub("[.]", "-", colnames(abundance_df))

  ## Plot
  column_annot <- assay(msfe_gsva@sfe_data[[id]], "es")["senScoreSenMayo",] %>%
    data.frame() %>%
    rename("SenMayo" = ".") %>%
    arrange(SenMayo)

  ## Order input column data by SenMayo enrichment score
  abundance_df <- abundance_df[,rownames(column_annot)]

  ## Order input row data by cell type
  abundance_df <- abundance_df[order(rownames(abundance_df)),]

  for (std in c("row", "column")) {
    ph <- pheatmap(abundance_df,
                   color = rev(cols4all::c4a(palette = "brewer.br_bg", n = 9)),
                   show_rownames = TRUE,
                   show_colnames = FALSE,
                   cluster_rows = FALSE,
                   cluster_cols = TRUE,
                   main = "Cell Type Density",
                   scale = std,
                   annotation_col = column_annot,
                   labels_row = names(selectedCellTypes)[order(names(selectedCellTypes))]
                   )

    prfx <- id
    main <- "_CellAbundance"
    sfx <- "_heatmap"
    other <- ifelse(std == "row", "_row_std", "_col_std")

    ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
           plot = ph,
           device = "png",
           width = 8.86,
           height = 7.43,
           units = "in",
           dpi = 300)
  }

  ## Housekeeping
  rm(abundance_df)
}


# ---------------------------------------------------------------------------- #
# ----Re-load images -----------------------------------------------------------
for (id in sampleNames_selected) {
  sfe <- getSFE(msfe, id)
  sfe@int_metadata$imgData <- NULL
  ## Get scale factors
  scaleF <- jsonlite::fromJSON(txt = file.path(sampleDir_selected[[id]],
                                               "outs/spatial",
                                               "scalefactors_json.json"))
  sfe <- SpatialFeatureExperiment::addImg(sfe,
                                          file.path(sampleDir_selected[[id]],
                                                    "outs/spatial/tissue_lowres_image.png"),
                                          sample_id = id,
                                          image_id = "lowres",
                                          scale_fct = scaleF[["tissue_lowres_scalef"]])
  sfe <- SpatialFeatureExperiment::mirrorImg(sfe,
                                             sample_id = id,
                                             image_id = "lowres")
  msfe <- addSFE(msfe, sfe, id)
  ## Housekeeping
  rm(sfe)
}



# ---------------------------------------------------------------------------- #
# ----Maybe to Remove later ----------------------------------------------------
### ----Calculate library size factors -----------------------------------------
## Calculate library size factors
msfe <- computeLibSizeFactors(msfe)

## Create folder
graphics_out <- paste0(projectFolder, "/graphics_out/Normalisation/")
dir.create(graphics_out, recursive = TRUE)

for (id in sampleNames_selected) {
  message("Working on sample: ", id)

  ## Extract sample
  sfe <- getSFE(msfe, sample_id = id)

  ## Density and histogram of library sizes
  p1 <- plotQC_sizeFactors(sfe)

  ## Map library sizes
  p2 <- plotQC_map(sfe,
                   metric = "custom",
                   metric_name = "sizeFactor",
                   metric_lab = "Size factor",
                   type = "hex")

  ## Patch, print and save
  print(p1 + p2)

  prfx <- id
  main <- "_SizeFactors"
  sfx <- ""
  other <- ""

  ggsave(paste0(graphics_out, prfx, main, sfx, other, ".png"),
         device = "png",
         width = grDevices::dev.size(units = "in")[1],
         height = grDevices::dev.size(units = "in")[2],
         units = "in",
         dpi = 300)

  ## Update msfe object
  msfe@sfe_data[[id]] <- sfe

  ## Housekeeping
  rm(sfe, p1, p2)
}


### ----Log-normalisation ------------------------------------------------------
## Calculate logcounts using library size factors
msfe <- normaliseCounts(msfe)


# ---------------------------------------------------------------------------- #
### ----Log-transformation -----------------------------------------------------
## Calculate logcounts without library normalisation
msfe <- normaliseCounts(msfe, method = "log2_t")
